<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Integrating Single-Cell Results for Exploring and Analyzing Methylation â€¢ iscream</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="lightswitch.js"></script><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><script src="deps/MathJax-3.2.2/tex-chtml.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Integrating Single-Cell Results for Exploring and Analyzing Methylation">
<meta name="description" content="Fast and memory efficient methylation bed file reader and analyzer. It can be used with both bulk and scWGBS data.">
<meta property="og:description" content="Fast and memory efficient methylation bed file reader and analyzer. It can be used with both bulk and scWGBS data.">
<meta property="og:image" content="https://huishenlab.github.io/iscream/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">iscream</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/huishenlab/iscream/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level2">
<h2 id="iscream-">iscream <a href="https://huishenlab.github.io/iscream/"><img src="reference/figures/logo.png" align="right" height="138" style="float:right; height:138px;"></a>
<a class="anchor" aria-label="anchor" href="#iscream-"></a>
</h2>
<p><em>Integrating Single-Cell Results for Exploring and Analyzing Methylation</em></p>
<!-- badges: start -->

<p>Analysis and visualization of Whole Genome Bisulfite Sequencing (WGBS) data requires reading aligned sequencing data into formats that existing packages like <a href="https://bioconductor.org/packages/devel/bioc/html/bsseq.html" class="external-link">BSseq</a> and <a href="https://github.com/andreaskapou/scMET" class="external-link">scMET</a> can analyze. Getting the data from on-disk formats like bedfiles to a matrix of methylation values can be difficult because, with nearly 30 million CpGs, WGBS data can be quite large.</p>
<p>iscream aims to efficiently read alinged (sc)WGBS data into formats that can be used by other packages. iscream uses <a href="https://www.htslib.org/" class="external-link">htslib</a> to query genomic regions to make matrices for BSSeq or aggregate the methylated reads for scMET.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="system-dependencies">System dependencies<a class="anchor" aria-label="anchor" href="#system-dependencies"></a>
</h3>
<div class="section level4">
<h4 id="htslib">
<em>htslib</em><a class="anchor" aria-label="anchor" href="#htslib"></a>
</h4>
<p><em>iscream</em> depends on the <em>htslib</em> header files. These may be installed with your package manager:</p>
<ul>
<li>ubuntu/debian: <code>libhts-dev</code><br>
</li>
<li>fedora/RHEL: <code>htslib-devel</code><br>
</li>
<li>brew: <code>htslib</code><br>
</li>
<li>nixpkgs: <code>htslib</code>
</li>
</ul>
<p>They may also be found among your HPC modules - make sure the <code>PKG_CONFIG_PATH</code> environment variable includes the <code>pkgconfig</code> location for your installation of <em>htslib</em>. You can verify that the <em>htslib</em> development libraries are installed with <code>pkg-config</code>:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">pkg-config</span> <span class="at">--cflags</span> <span class="at">--libs</span> htslib</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="gcc">GCC<a class="anchor" aria-label="anchor" href="#gcc"></a>
</h4>
<p>GNU GCC must be installed for OpenMP support. This is usually installed by default on Linux systems, but may need to be manually installed on MacOS to use iscream with multiple threads<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;Using OpenMP is also possible with Clang on MacOS (&lt;a href="https://mac.r-project.org/openmp/" class="external-link uri"&gt;https://mac.r-project.org/openmp/&lt;/a&gt;) but installing GCC with Homebrew may be easier (&lt;a href="https://formulae.brew.sh/formula/gcc" class="external-link uri"&gt;https://formulae.brew.sh/formula/gcc&lt;/a&gt;).&lt;/p&gt;'><sup>1</sup></a>.</p>
</div>
</div>
<div class="section level3">
<h3 id="github">GitHub<a class="anchor" aria-label="anchor" href="#github"></a>
</h3>
<p>You can install the development version from Github by cloning the repo and running</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">git</span> clone https://github.com/huishenlab/iscream</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="ex">R</span> CMD INSTALL iscream</span></code></pre></div>
<p>You can also use the R <a href="https://devtools.r-lib.org/" class="external-link"><code>devtools</code></a> package:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"huishenlab/iscream"</span><span class="op">)</span></span></code></pre></div>
<p>or <a href="https://pak.r-lib.org/" class="external-link"><code>pak</code></a>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pak</span><span class="fu">::</span><span class="fu"><a href="https://pak.r-lib.org/reference/pkg_install.html" class="external-link">pkg_install</a></span><span class="op">(</span><span class="st">"huishenlab/iscream"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h3>
<p>A user guide is available on the <a href="https://huishenlab.github.io/iscream/">package website</a>. Bug reports may be submitted through GitHub issues.</p>
</div>
</div>


  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/huishenlab/iscream/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/huishenlab/iscream/issues/" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing iscream</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>James Eapen <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-6016-3598" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Jacob Morrison <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-8592-4744" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Hui Shen <br><small class="roles"> Thesis advisor, funder </small> <a href="https://orcid.org/0000-0001-9767-4084" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/huishenlab/iscream/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/huishenlab/iscream/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://app.codecov.io/gh/huishenlab/iscream" class="external-link"><img src="https://codecov.io/gh/huishenlab/iscream/graph/badge.svg" alt="Codecov test coverage"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by James Eapen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
