---
title: "An introduction to iscream"
author:
- name: James Eapen
  affiliation:
  - Department of Epigenetics, Van Andel Institute, Grand Rapids, MI
output:
  BiocStyle::html_document:
    toc_float: true
vignette: >
  %\VignetteIndexEntry{An introduction to iscream}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: refs.bib
link-citations: yes
---

```{r setup, include=FALSE}
options("datatable.print.nrows" = 50)
library(BiocStyle)
```

## Introduction

iscream is a BED file querying package that can can retrieve records from
multiple [tabixed](https://en.wikipedia.org/wiki/Tabix) BED files based on a
input regions of interest. It can make queries like tabix, summarize data from
loci within the regions of interest and make matrices across input files and
regions. All operations can be done in parallel and return R/Bioconductor data
structures.

iscream was designed to be memory efficient and fast. For more information on
performance considerations on large datasets see `vignette("performance")`.

### Loading iscream

```{r load_iscream}
library(iscream)
set_threads(2)
```

On load, iscream will inform the user about the number if threads it is set to
use. This is configurable with either `set_threads()` or by setting
`option("iscream.threads")`, which can be done in your `.Rprofile` to set it on
startup. Once loaded you use the following querying functions.

## `tabix`

tabix is a command-line tool that queries loci from compressed BED files. As
input it receives the path to a tabixed BED file and one or more genomic regions
of interest. The regions of interest may be passed in the `"chr:start-end"`
format or as a BED file. Given these inputs, it prints the loci from the tabixed
BED file that fall within the regions of interest. This output may be redirected
to a file and read into R.

iscream's `tabix()` function works in a similar way but supports multiple files
and multiple input formats all in one function. It receives a vector of paths to
tabixed BED files and regions of interest. The input regions may be a vector of
`"chr:start-end"` strings, a data frame with `"chr"`, `"start"`, and `"end"`
columns or a `GRanges` object. It queries the BEDfiles and returns the loci as a
parsed `r CRANpkg("data.table")`, or a `GRanges` object if the input is a
`GRanges` object.

Using iscream eliminates context switching between the shell and R for tabix
queries since the queries are made in R and the result is an R data structure.
Rsamtools is another R package that can query data from BED files, but it only
supports one file at a time and only accepts `GRanges` as the input regions. See
`vignette("tabix")` for a comparison of iscream and Rsamtools.

### Setup

Using `list.files()`, you can get the paths to all compressed bed files in a
directory. These files contain small regions from chromosome 1 and Y from the
snmC-seq2 methylation data data [@luo2018a].

```{r load_data}
data_dir <- system.file("extdata", package = "iscream")
(bedfiles <- list.files(
  data_dir,
  pattern = "cell[1-5].bed.gz$",
  full.names = TRUE))
```

For demonstration, I'm using two regions.

```{r make_regions}
regions <- c("chr1:184577-680065", "chrY:56877780-56882524")
```

**Note**: in this dataset the chromosome identifier is in a "chr[number]"
format. However, in some datasets, the chromosome ID is just a number or a
letter like '1', or 'Y'. To see the chromosome ID format across your files you
can use `query_chroms()` to get all unique chromosomes across all input files:

```{r query_chroms}
query_chroms(bedfiles)
```

### Making queries

`tabix()` one file:

```{r tabix_one}
tabix(bedfiles[1], regions)
```

With multiple files, the output contains a column for the file name.

```{r tabix_all}
tabix(bedfiles, regions)
```

### Using `GRanges`

To get `r Biocpkg("GenomicRanges")` objects, use a `GRanges` object as the input regions, or call
`GRanges()` on the output. With `GRanges` input, the output `ranges` are
computed with `subsetByOverlaps`.

```{r tabix_gr, message=FALSE}
library(GenomicRanges)
gr <- GRanges(regions)
tabix(bedfiles, gr) |> makeGRangesFromDataFrame(keep.extra.columns = TRUE)
```

### Setting column names

You can set the result data column names or the `GRanges` `mcols` with
`col.names`:

```{r tabix_colnames}
tabix(bedfiles, regions, col.names = c("beta", "coverage"))
tabix(bedfiles, gr, col.names = c("beta", "coverage")) |> makeGRangesFromDataFrame(keep.extra.columns = TRUE)
```

### Rsamtools-style output

`tabix()` also has a `raw` argument if you want an unparsed named list like
Rsamtools, but with multi-file support:

```{r tabix_raw}
tabix(bedfiles, regions, raw = TRUE)
```

### WGBS BED files

Since iscream was originally developed to read Whole Genome Bisulfite Sequencing
data, it has the `aligner` argument for automatically setting column names for
the BISCUIT [@zhou2024], Bismark [@krueger2011], and BSBolt [@farrell2021].

## `summarize_regions`

Using the same bedfiles and regions, iscream can also summarize the data per
region. This requires providing the index of the data columns you want
summarized and their column names. By default, all summarizing functions are run
on each input column:

- Sum: `"sum"`
- Mean: `"mean"`
- Median: `"median"`
- Standard deviation: `"stddev"`
- Variance: `"variance"`
- Minimum: `"min"`
- Maximum: `"max"`
- Range: `"range"`
- No. of records in the region: `"count"`

```{r summary}
summarize_regions(
  bedfiles,
  regions,
  columns = c(4, 5),
  col_names = c("beta", "coverage")
)
```

The `feature` column here contains the genomic region coordinates, but can be
set to something more informational if you have named regions. You can also
select function with `fun`:


```{r summary_named}
names(regions) <- c("R1", "R2")
summarize_regions(
  bedfiles,
  regions,
  fun = c("mean", "sum"),
  columns = 5,
  col_names = "coverage"
)
```

### WGBS BED files

For WGBS data specifically, you can use `summarize_meth_regions()`, which takes
the `aligner` argument to correctly parse the columns.

```{r summarize_meth_regions}
summarize_meth_regions(
  bedfiles,
  regions,
  fun = c("mean", "sum")
)
```


## `make_mat`

`make_mat` makes a matrix where every row is a locus found in at least one input
file and the columns are the input files. It returns a `r Biocpkg("RangedSummarizedExperiment")`
by default, but can also return a `GRanges` object or a named list of the
matrix, coordinates, and file names.

```{r make_mat, message=FALSE}
library(SummarizedExperiment)
mat <- make_mat(bedfiles, regions, column = 4, mat_name = "beta")
head(assay(mat), 10)
```

If you have sparse data, you can save memory with `sparse = TRUE`

```{r sparse}
mat <- make_mat(bedfiles, regions, column = 4, mat_name = "beta", sparse = TRUE)
head(assay(mat), 10)
```

### WGBS BED files

For a `r Biocpkg("BSseq")` compatible object, use `make_meth_mat` which returns
a list of BSseq inputs. To make the `BSseq` object run

```{r, eval=FALSE}
meth_mat <- make_mat(bedfiles, regions)
do.call(BSseq, meth_mat)
```

For more information on the supported Bioconductor data structures see
`r Biocpkg("iscream", "data_structures")`. An example workflow using
iscream to examine transcription start site methylation is at
`r Biocpkg("iscream", "TSS")`.

## Session info

```{r si}
sessionInfo()
```

## References
