<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Getting other data structures from iscream • iscream</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="../deps/MathJax-3.2.2/tex-chtml.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting other data structures from iscream">
<meta name="robots" content="noindex">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">iscream</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">0.1.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/iscream.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/data_structures.html">Getting other data structures from iscream</a></li>
    <li><a class="dropdown-item" href="../articles/TSS.html">Plotting TSS methylation profiles</a></li>
    <li><a class="dropdown-item" href="../articles/htslib.html">Getting htslib headers</a></li>
    <li><a class="dropdown-item" href="../articles/manuscript_data.html">Manuscript data availabiliy</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/huishenlab/iscream/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Getting other data structures from iscream</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/huishenlab/iscream/blob/dev/vignettes/data_structures.Rmd" class="external-link"><code>vignettes/data_structures.Rmd</code></a></small>
      <div class="d-none name"><code>data_structures.Rmd</code></div>
    </div>

    
    
<p>The examples here show how iscream output can be converted into other
data structures for further analysis.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://huishenlab.github.io/iscream/">iscream</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## iscream using 1 thread by default but parallelly::availableCores() detects 4 possibly available threads. See `?set_threads` for information on multithreading before trying to use more.</span></span>
<span><span class="co">## 'tabix' executable not found in $PATH. tabix() will use htslib to make queries instead which can be slower. See ?tabix for details.</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"iscream"</span><span class="op">)</span></span>
<span><span class="va">bedfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">data_dir</span>, pattern <span class="op">=</span> <span class="st">"[a|b|c|d].bed.gz$"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>A <span class="op">=</span> <span class="st">"chr1:1-6"</span>, B <span class="op">=</span> <span class="st">"chr1:7-10"</span>, C <span class="op">=</span> <span class="st">"chr1:11-14"</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="granges">GRanges<a class="anchor" aria-label="anchor" href="#granges"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span></code></pre></div>
<p>GRanges can be used as the input to all of iscream’s querying
functions. The output of <code><a href="../reference/tabix.html">tabix()</a></code> and
<code><a href="../reference/make_mat.html">make_mat()</a></code>, <code><a href="../reference/make_bsseq_mat.html">make_bsseq_mat()</a></code> can also be
turned into GRanges.</p>
<div class="section level3">
<h3 id="from-tabix-queries">From <code>tabix</code> queries<a class="anchor" aria-label="anchor" href="#from-tabix-queries"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tabix.html">tabix</a></span><span class="op">(</span><span class="va">bedfiles</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">regions</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 7 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |        V4        V5</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">##   [1]     chr1       0-2      * |       1.0         1</span></span>
<span><span class="co">##   [2]     chr1       2-4      * |       1.0         1</span></span>
<span><span class="co">##   [3]     chr1       4-6      * |       0.0         2</span></span>
<span><span class="co">##   [4]     chr1       6-8      * |       0.0         1</span></span>
<span><span class="co">##   [5]     chr1      8-10      * |       0.5         2</span></span>
<span><span class="co">##   [6]     chr1     10-12      * |       1.0         2</span></span>
<span><span class="co">##   [7]     chr1     12-14      * |       1.0         3</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>If the input is already a <code>GRanges</code> object,
<code><a href="../reference/tabix.html">tabix()</a></code> will also return a <code>GRanges</code> object
along with any attached metadata. <code><a href="../reference/tabix.html">tabix()</a></code> uses the input
<code><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">findOverlaps()</a></code> to join input metadata with the queried
data.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="va">regions</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">values</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></span><span class="op">(</span></span>
<span>    gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span>, <span class="st">"gene3"</span><span class="op">)</span>,</span>
<span>    some_metadata <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"s1"</span>, <span class="st">"s2"</span>, <span class="st">"s3"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">gr</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 2 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand |        gene some_metadata</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;   &lt;character&gt;</span></span>
<span><span class="co">##   A     chr1       1-6      * |       gene1            s1</span></span>
<span><span class="co">##   B     chr1      7-10      * |       gene2            s2</span></span>
<span><span class="co">##   C     chr1     11-14      * |       gene3            s3</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tabix.html">tabix</a></span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">gr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 20 ranges and 5 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        V4        V5        file        gene</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##    [1]     chr1       1-2      * |       1.0         1           a       gene1</span></span>
<span><span class="co">##    [2]     chr1       3-4      * |       1.0         1           a       gene1</span></span>
<span><span class="co">##    [3]     chr1       5-6      * |       0.0         2           a       gene1</span></span>
<span><span class="co">##    [4]     chr1       7-8      * |       0.0         1           a       gene2</span></span>
<span><span class="co">##    [5]     chr1      9-10      * |       0.5         2           a       gene2</span></span>
<span><span class="co">##    ...      ...       ...    ... .       ...       ...         ...         ...</span></span>
<span><span class="co">##   [16]     chr1       1-2      * |       1.0         1           d       gene1</span></span>
<span><span class="co">##   [17]     chr1       3-4      * |       1.0         2           d       gene1</span></span>
<span><span class="co">##   [18]     chr1       7-8      * |       0.0         1           d       gene2</span></span>
<span><span class="co">##   [19]     chr1      9-10      * |       0.5         2           d       gene2</span></span>
<span><span class="co">##   [20]     chr1     13-14      * |       1.0         1           d       gene3</span></span>
<span><span class="co">##        some_metadata</span></span>
<span><span class="co">##          &lt;character&gt;</span></span>
<span><span class="co">##    [1]            s1</span></span>
<span><span class="co">##    [2]            s1</span></span>
<span><span class="co">##    [3]            s1</span></span>
<span><span class="co">##    [4]            s2</span></span>
<span><span class="co">##    [5]            s2</span></span>
<span><span class="co">##    ...           ...</span></span>
<span><span class="co">##   [16]            s1</span></span>
<span><span class="co">##   [17]            s1</span></span>
<span><span class="co">##   [18]            s2</span></span>
<span><span class="co">##   [19]            s2</span></span>
<span><span class="co">##   [20]            s3</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>If the input BED file is not zero-based (e.g. Bismark coverage
files), set <code>zero_based = FALSE</code> in the <code><a href="../reference/tabix.html">tabix()</a></code>
call to get the correct conversion from data frame to GenomicRanges.</p>
</div>
<div class="section level3">
<h3 id="from-summarize_regions">From <code>summarize_regions</code><a class="anchor" aria-label="anchor" href="#from-summarize_regions"></a>
</h3>
<p>With a named set of regions, you’ll need to
<code>set_region_rownames</code> as the input to create the GRanges
object:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">summary</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_meth_regions.html">summarize_meth_regions</a></span><span class="op">(</span></span>
<span>  <span class="va">bedfiles</span>,</span>
<span>  <span class="va">regions</span>,</span>
<span>  set_region_rownames <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  fun <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sum"</span>, <span class="st">"mean"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [19:21:25.184940] [iscream::summarize_regions] [info] Summarizing 3 regions from 4 bedfiles</span></span>
<span><span class="co">## [19:21:25.184989] [iscream::summarize_regions] [info] using sum, mean</span></span>
<span><span class="co">## [19:21:25.184993] [iscream::summarize_regions] [info] with columns 4, 5 as coverage, M</span></span></code></pre>
<pre><code><span><span class="co">##            Feature Sample coverage.sum M.sum coverage.mean    M.mean</span></span>
<span><span class="co">## chr1:1-6         A      a            4     2      1.333333 0.6666667</span></span>
<span><span class="co">## chr1:7-10        B      a            3     1      1.500000 0.5000000</span></span>
<span><span class="co">## chr1:11-14       C      a            5     5      2.500000 2.5000000</span></span>
<span><span class="co">## chr1:1-6         A      b            4     2      2.000000 1.0000000</span></span>
<span><span class="co">## chr1:7-10        B      b            1     1      1.000000 1.0000000</span></span>
<span><span class="co">## chr1:11-14       C      b            3     1      1.500000 0.5000000</span></span>
<span><span class="co">## chr1:1-6         A      c            2     2      2.000000 2.0000000</span></span>
<span><span class="co">## chr1:7-10        B      c            3     1      1.500000 0.5000000</span></span>
<span><span class="co">## chr1:11-14       C      c           NA    NA            NA        NA</span></span>
<span><span class="co">## chr1:1-6         A      d            3     3      1.500000 1.5000000</span></span>
<span><span class="co">## chr1:7-10        B      d            3     1      1.500000 0.5000000</span></span>
<span><span class="co">## chr1:11-14       C      d            1     1      1.000000 1.0000000</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">summary</span><span class="op">)</span>, summary <span class="op">=</span> <span class="va">summary</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 12 ranges and 6 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand | summary.Feature summary.Sample</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;    &lt;character&gt;</span></span>
<span><span class="co">##    [1]     chr1       1-6      * |               A              a</span></span>
<span><span class="co">##    [2]     chr1      7-10      * |               B              a</span></span>
<span><span class="co">##    [3]     chr1     11-14      * |               C              a</span></span>
<span><span class="co">##    [4]     chr1       1-6      * |               A              b</span></span>
<span><span class="co">##    [5]     chr1      7-10      * |               B              b</span></span>
<span><span class="co">##    ...      ...       ...    ... .             ...            ...</span></span>
<span><span class="co">##    [8]     chr1      7-10      * |               B              c</span></span>
<span><span class="co">##    [9]     chr1     11-14      * |               C              c</span></span>
<span><span class="co">##   [10]     chr1       1-6      * |               A              d</span></span>
<span><span class="co">##   [11]     chr1      7-10      * |               B              d</span></span>
<span><span class="co">##   [12]     chr1     11-14      * |               C              d</span></span>
<span><span class="co">##        summary.coverage.sum summary.M.sum summary.coverage.mean summary.M.mean</span></span>
<span><span class="co">##                   &lt;numeric&gt;     &lt;numeric&gt;             &lt;numeric&gt;      &lt;numeric&gt;</span></span>
<span><span class="co">##    [1]                    4             2               1.33333       0.666667</span></span>
<span><span class="co">##    [2]                    3             1               1.50000       0.500000</span></span>
<span><span class="co">##    [3]                    5             5               2.50000       2.500000</span></span>
<span><span class="co">##    [4]                    4             2               2.00000       1.000000</span></span>
<span><span class="co">##    [5]                    1             1               1.00000       1.000000</span></span>
<span><span class="co">##    ...                  ...           ...                   ...            ...</span></span>
<span><span class="co">##    [8]                    3             1                   1.5            0.5</span></span>
<span><span class="co">##    [9]                   NA            NA                    NA             NA</span></span>
<span><span class="co">##   [10]                    3             3                   1.5            1.5</span></span>
<span><span class="co">##   [11]                    3             1                   1.5            0.5</span></span>
<span><span class="co">##   [12]                    1             1                   1.0            1.0</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>If the regions vector is not named, pass the <code>Feature</code>
column as the GRanges input regions:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">summary</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_meth_regions.html">summarize_meth_regions</a></span><span class="op">(</span></span>
<span>  <span class="va">bedfiles</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unname.html" class="external-link">unname</a></span><span class="op">(</span><span class="va">regions</span><span class="op">)</span>,</span>
<span>  fun <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sum"</span>, <span class="st">"mean"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [19:21:25.302993] [iscream::summarize_regions] [info] Summarizing 3 regions from 4 bedfiles</span></span>
<span><span class="co">## [19:21:25.303018] [iscream::summarize_regions] [info] using sum, mean</span></span>
<span><span class="co">## [19:21:25.303022] [iscream::summarize_regions] [info] with columns 4, 5 as coverage, M</span></span></code></pre>
<pre><code><span><span class="co">##       Feature Sample coverage.sum M.sum coverage.mean    M.mean</span></span>
<span><span class="co">## 1    chr1:1-6      a            4     2      1.333333 0.6666667</span></span>
<span><span class="co">## 2   chr1:7-10      a            3     1      1.500000 0.5000000</span></span>
<span><span class="co">## 3  chr1:11-14      a            5     5      2.500000 2.5000000</span></span>
<span><span class="co">## 4    chr1:1-6      b            4     2      2.000000 1.0000000</span></span>
<span><span class="co">## 5   chr1:7-10      b            1     1      1.000000 1.0000000</span></span>
<span><span class="co">## 6  chr1:11-14      b            3     1      1.500000 0.5000000</span></span>
<span><span class="co">## 7    chr1:1-6      c            2     2      2.000000 2.0000000</span></span>
<span><span class="co">## 8   chr1:7-10      c            3     1      1.500000 0.5000000</span></span>
<span><span class="co">## 9  chr1:11-14      c           NA    NA            NA        NA</span></span>
<span><span class="co">## 10   chr1:1-6      d            3     3      1.500000 1.5000000</span></span>
<span><span class="co">## 11  chr1:7-10      d            3     1      1.500000 0.5000000</span></span>
<span><span class="co">## 12 chr1:11-14      d            1     1      1.000000 1.0000000</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="va">summary</span><span class="op">$</span><span class="va">Feature</span>, summary <span class="op">=</span> <span class="va">summary</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 12 ranges and 5 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand | summary.Sample summary.coverage.sum</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;character&gt;            &lt;numeric&gt;</span></span>
<span><span class="co">##    [1]     chr1       1-6      * |              a                    4</span></span>
<span><span class="co">##    [2]     chr1      7-10      * |              a                    3</span></span>
<span><span class="co">##    [3]     chr1     11-14      * |              a                    5</span></span>
<span><span class="co">##    [4]     chr1       1-6      * |              b                    4</span></span>
<span><span class="co">##    [5]     chr1      7-10      * |              b                    1</span></span>
<span><span class="co">##    ...      ...       ...    ... .            ...                  ...</span></span>
<span><span class="co">##    [8]     chr1      7-10      * |              c                    3</span></span>
<span><span class="co">##    [9]     chr1     11-14      * |              c                   NA</span></span>
<span><span class="co">##   [10]     chr1       1-6      * |              d                    3</span></span>
<span><span class="co">##   [11]     chr1      7-10      * |              d                    3</span></span>
<span><span class="co">##   [12]     chr1     11-14      * |              d                    1</span></span>
<span><span class="co">##        summary.M.sum summary.coverage.mean summary.M.mean</span></span>
<span><span class="co">##            &lt;numeric&gt;             &lt;numeric&gt;      &lt;numeric&gt;</span></span>
<span><span class="co">##    [1]             2               1.33333       0.666667</span></span>
<span><span class="co">##    [2]             1               1.50000       0.500000</span></span>
<span><span class="co">##    [3]             5               2.50000       2.500000</span></span>
<span><span class="co">##    [4]             2               2.00000       1.000000</span></span>
<span><span class="co">##    [5]             1               1.00000       1.000000</span></span>
<span><span class="co">##    ...           ...                   ...            ...</span></span>
<span><span class="co">##    [8]             1                   1.5            0.5</span></span>
<span><span class="co">##    [9]            NA                    NA             NA</span></span>
<span><span class="co">##   [10]             3                   1.5            1.5</span></span>
<span><span class="co">##   [11]             1                   1.5            0.5</span></span>
<span><span class="co">##   [12]             1                   1.0            1.0</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="making-summarizedexperiment-objects">Making SummarizedExperiment objects<a class="anchor" aria-label="anchor" href="#making-summarizedexperiment-objects"></a>
</h2>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">"SummarizedExperiment"</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: replacing previous import 'S4Arrays::makeNindexFromArrayViewport' by</span></span>
<span><span class="co">## 'DelayedArray::makeNindexFromArrayViewport' when loading 'SummarizedExperiment'</span></span></code></pre>
<div class="section level3">
<h3 id="from-make_mat-or-make_bsseq_mat-matrices">From <code>make_mat()</code> or <code>make_bsseq_mat()</code>
matrices<a class="anchor" aria-label="anchor" href="#from-make_mat-or-make_bsseq_mat-matrices"></a>
</h3>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">mats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_bsseq_mat.html">make_bsseq_mat</a></span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [19:21:28.354612] [iscream::query_all] [info] Querying 3 regions from 4 bedfiles</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [19:21:28.354974] [iscream::query_all] [info] Creating metadata vectors</span></span>
<span><span class="co">## [19:21:28.355013] [iscream::query_all] [info] 7 loci found - 9993 extra rows allocated with 0 resizes</span></span>
<span><span class="co">## [19:21:28.355017] [iscream::query_all] [info] Creating dense matrix</span></span></code></pre>
<pre><code><span><span class="co">## $M</span></span>
<span><span class="co">##      a b c d</span></span>
<span><span class="co">## [1,] 1 0 0 1</span></span>
<span><span class="co">## [2,] 1 0 2 2</span></span>
<span><span class="co">## [3,] 0 2 0 0</span></span>
<span><span class="co">## [4,] 0 1 0 0</span></span>
<span><span class="co">## [5,] 1 0 1 1</span></span>
<span><span class="co">## [6,] 2 0 0 0</span></span>
<span><span class="co">## [7,] 3 1 0 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $Cov</span></span>
<span><span class="co">##      a b c d</span></span>
<span><span class="co">## [1,] 1 2 0 1</span></span>
<span><span class="co">## [2,] 1 0 2 2</span></span>
<span><span class="co">## [3,] 2 2 0 0</span></span>
<span><span class="co">## [4,] 1 1 2 1</span></span>
<span><span class="co">## [5,] 2 0 1 2</span></span>
<span><span class="co">## [6,] 2 2 0 0</span></span>
<span><span class="co">## [7,] 3 1 0 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $pos</span></span>
<span><span class="co">## [1]  0  2  4  6  8 10 12</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $chr</span></span>
<span><span class="co">## [1] "chr1" "chr1" "chr1" "chr1" "chr1" "chr1" "chr1"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $sampleNames</span></span>
<span><span class="co">## [1] "a" "b" "c" "d"</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">mats.sparse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_bsseq_mat.html">make_bsseq_mat</a></span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span>, sparse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [19:21:28.356082] [iscream::query_all] [info] Querying 3 regions from 4 bedfiles</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [19:21:28.356404] [iscream::query_all] [info] Creating metadata vectors</span></span>
<span><span class="co">## [19:21:28.356423] [iscream::query_all] [info] 7 loci found - 9993 extra rows allocated with 0 resizes</span></span>
<span><span class="co">## [19:21:28.356432] [iscream::query_all] [info] Creating sparse matrix</span></span></code></pre>
<pre><code><span><span class="co">## $M</span></span>
<span><span class="co">## 7 x 4 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##      a b c d</span></span>
<span><span class="co">## [1,] 1 . . 1</span></span>
<span><span class="co">## [2,] 1 . 2 2</span></span>
<span><span class="co">## [3,] . 2 . .</span></span>
<span><span class="co">## [4,] . 1 . .</span></span>
<span><span class="co">## [5,] 1 . 1 1</span></span>
<span><span class="co">## [6,] 2 . . .</span></span>
<span><span class="co">## [7,] 3 1 . 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $Cov</span></span>
<span><span class="co">## 7 x 4 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##      a b c d</span></span>
<span><span class="co">## [1,] 1 2 . 1</span></span>
<span><span class="co">## [2,] 1 . 2 2</span></span>
<span><span class="co">## [3,] 2 2 . .</span></span>
<span><span class="co">## [4,] 1 1 2 1</span></span>
<span><span class="co">## [5,] 2 . 1 2</span></span>
<span><span class="co">## [6,] 2 2 . .</span></span>
<span><span class="co">## [7,] 3 1 . 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $pos</span></span>
<span><span class="co">## [1]  0  2  4  6  8 10 12</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $chr</span></span>
<span><span class="co">## [1] "chr1" "chr1" "chr1" "chr1" "chr1" "chr1" "chr1"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $sampleNames</span></span>
<span><span class="co">## [1] "a" "b" "c" "d"</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="va">mats</span><span class="op">$</span><span class="va">chr</span>, <span class="va">mats</span><span class="op">$</span><span class="va">pos</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># dense</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>M <span class="op">=</span> <span class="va">mats</span><span class="op">$</span><span class="va">M</span>, Cov <span class="op">=</span> <span class="va">mats</span><span class="op">$</span><span class="va">Cov</span><span class="op">)</span>, rowRanges <span class="op">=</span> <span class="va">gr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 7 4 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(2): M Cov</span></span>
<span><span class="co">## rownames: NULL</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(4): a b c d</span></span>
<span><span class="co">## colData names(0):</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># sparse</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></span><span class="op">(</span></span>
<span>  assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>M <span class="op">=</span> <span class="va">mats.sparse</span><span class="op">$</span><span class="va">M</span>, Cov <span class="op">=</span> <span class="va">mats.sparse</span><span class="op">$</span><span class="va">Cov</span><span class="op">)</span>,</span>
<span>  rowRanges <span class="op">=</span> <span class="va">gr</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 7 4 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(2): M Cov</span></span>
<span><span class="co">## rownames: NULL</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(4): a b c d</span></span>
<span><span class="co">## colData names(0):</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="making-bsseq-objects">Making <code>BSseq</code> objects<a class="anchor" aria-label="anchor" href="#making-bsseq-objects"></a>
</h2>
<p>A BSseq object is a type of SummarizedExperiment, but it cannot
handle sparse matrices:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bsseq</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">BSseq</span>, <span class="va">mats</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.0 (2025-04-11)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] SummarizedExperiment_1.38.0 Biobase_2.68.0             </span></span>
<span><span class="co">##  [3] MatrixGenerics_1.20.0       matrixStats_1.5.0          </span></span>
<span><span class="co">##  [5] GenomicRanges_1.60.0        GenomeInfoDb_1.44.0        </span></span>
<span><span class="co">##  [7] IRanges_2.42.0              S4Vectors_0.46.0           </span></span>
<span><span class="co">##  [9] BiocGenerics_0.54.0         generics_0.1.3             </span></span>
<span><span class="co">## [11] iscream_0.1.0.9000         </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] sass_0.4.10             SparseArray_1.8.0       lattice_0.22-6         </span></span>
<span><span class="co">##  [4] digest_0.6.37           evaluate_1.0.3          grid_4.5.0             </span></span>
<span><span class="co">##  [7] fastmap_1.2.0           jsonlite_2.0.0          Matrix_1.7-3           </span></span>
<span><span class="co">## [10] httr_1.4.7              UCSC.utils_1.4.0        textshaping_1.0.0      </span></span>
<span><span class="co">## [13] jquerylib_0.1.4         abind_1.4-8             cli_3.6.5              </span></span>
<span><span class="co">## [16] rlang_1.1.6             crayon_1.5.3            XVector_0.48.0         </span></span>
<span><span class="co">## [19] parallelly_1.43.0       cachem_1.1.0            DelayedArray_0.34.1    </span></span>
<span><span class="co">## [22] yaml_2.3.10             S4Arrays_1.8.0          tools_4.5.0            </span></span>
<span><span class="co">## [25] parallel_4.5.0          GenomeInfoDbData_1.2.14 R6_2.6.1               </span></span>
<span><span class="co">## [28] lifecycle_1.0.4         fs_1.6.6                stringfish_0.16.0      </span></span>
<span><span class="co">## [31] ragg_1.4.0              desc_1.4.3              pkgdown_2.1.1          </span></span>
<span><span class="co">## [34] RcppParallel_5.1.10     bslib_0.9.0             data.table_1.17.0      </span></span>
<span><span class="co">## [37] Rcpp_1.0.14             systemfonts_1.2.2       xfun_0.52              </span></span>
<span><span class="co">## [40] knitr_1.50              htmltools_0.5.8.1       rmarkdown_2.29         </span></span>
<span><span class="co">## [43] compiler_4.5.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by James Eapen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
