<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>An introduction to iscream • iscream</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="An introduction to iscream">
<meta name="robots" content="noindex">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">iscream</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">0.99.10</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/iscream.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/performance.html">Improving iscream performance</a></li>
    <li><a class="dropdown-item" href="../articles/data_structures.html">iscream compatible data structures</a></li>
    <li><a class="dropdown-item" href="../articles/TSS.html">Plotting TSS methylation profiles</a></li>
    <li><a class="dropdown-item" href="../articles/tabix.html">iscream vs Rsamtools::scanTabix</a></li>
    <li><a class="dropdown-item" href="../articles/htslib.html">Getting htslib headers</a></li>
    <li><a class="dropdown-item" href="../articles/manuscript_data.html">Manuscript data availabiliy</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/huishenlab/iscream/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>An introduction to iscream</h1>
                        <h4 data-toc-skip class="author">James
Eapen</h4>
            <address class="author_afil">
      Department of Epigenetics, Van Andel Institute, Grand Rapids,
MI<br><small class="dont-index">Source: <a href="https://github.com/huishenlab/iscream/blob/dev/vignettes/iscream.Rmd" class="external-link"><code>vignettes/iscream.Rmd</code></a></small>
      <div class="d-none name"><code>iscream.Rmd</code></div>
    </address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>iscream is a BED file querying package that can can retrieve records
within regions of interest from multiple <a href="https://en.wikipedia.org/wiki/Tabix" class="external-link">tabixed</a> BED files. It can
make queries like tabix, summarize data within regions and make matrices
across input files and regions. All operations can be done in parallel
and return R/Bioconductor data structures.</p>
<p>iscream was designed to be memory efficient and fast on large
datasets. This vignette demonstrates iscream’s functions on a small
dataset. For more information on performance and considerations on
larger real datasets see <code><a href="../articles/performance.html">vignette("performance")</a></code>.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>iscream may be installed from <a href="https://bioconductor.org" class="external-link uri">https://bioconductor.org</a> with</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"iscream"</span><span class="op">)</span></span></code></pre></div>
<p>Although iscream will compile using <code>Rhtslib</code>, it will
perform best if compiled against <a href="https://www.htslib.org/download/" class="external-link">htslib</a> which has itself been
compiled with <code>libdeflate</code>. See
<code><a href="../articles/htslib.html">vignette("htslib")</a></code> for more information.</p>
<div class="section level3">
<h3 id="loading-iscream">Loading iscream<a class="anchor" aria-label="anchor" href="#loading-iscream"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://huishenlab.github.io/iscream/">iscream</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## iscream using 1 thread by default but parallelly::availableCores() detects 4 possibly available threads. See `?set_threads` for information on multithreading before trying to use more.</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/set_threads.html">set_threads</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## iscream now using 2 of 4 available threads.</span></span></code></pre>
<p>On load, iscream will inform the user about the number if threads it
is set to use. This is configurable with either
<code><a href="../reference/set_threads.html">set_threads()</a></code> or by setting
<code>option("iscream.threads")</code>, which can be done in your
<code>.Rprofile</code> to set it on startup. Once loaded you use the
following querying functions.</p>
</div>
</div>
<div class="section level2">
<h2 id="tabix">
<code>tabix()</code><a class="anchor" aria-label="anchor" href="#tabix"></a>
</h2>
<p>tabix is a command-line tool that queries records from compressed BED
files. As input it receives the path to a tabixed BED file and one or
more genomic regions of interest. It queries and prints the records from
the tabixed BED file that fall within the regions of interest. This
output may be redirected to a file and read into R.</p>
<p>iscream’s <code><a href="../reference/tabix.html">tabix()</a></code> function works in a similar way but
supports multiple files and multiple input formats all in one function.
It receives a vector of paths to tabixed BED files and regions of
interest. The input regions may be a vector of
<code>"chr:start-end"</code> strings, a data frame with
<code>"chr"</code>, <code>"start"</code>, and <code>"end"</code> columns
or a <code>GRanges</code> object. It queries the BED files and returns
the records as a parsed <em><a href="https://CRAN.R-project.org/package=data.table" class="external-link">data.table</a></em>.
To get a <code>GRanges</code> object instead, use
<code>tabix_gr</code>.</p>
<p>Using iscream eliminates context switching between the shell and R
for tabix queries since the queries are made in R and the result is an R
data structure. <em><a href="https://bioconductor.org/packages/3.21/Rsamtools" class="external-link">Rsamtools</a></em>
is another R package that can query data from BED files, but it only
supports one file at a time, only accepts <code>GRanges</code> as the
input regions and does not parse the tab-delimited strings. See
<code><a href="../articles/tabix.html">vignette("tabix")</a></code> for a comparison of iscream and
Rsamtools.</p>
<div class="section level3">
<h3 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h3>
<p>Using <code><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files()</a></code>, you can get the paths to all
compressed bed files in a directory. These files contain small regions
from chromosome 1 and Y from the snmC-seq2 methylation data <span class="citation">(<a href="#ref-luo2018a">Luo et al.
2018</a>)</span>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"iscream"</span><span class="op">)</span></span>
<span><span class="op">(</span><span class="va">bedfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span></span>
<span>  <span class="va">data_dir</span>,</span>
<span>  pattern <span class="op">=</span> <span class="st">"cell[1-5].bed.gz$"</span>,</span>
<span>  full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "/home/runner/work/_temp/Library/iscream/extdata/cell1.bed.gz"</span></span>
<span><span class="co">## [2] "/home/runner/work/_temp/Library/iscream/extdata/cell2.bed.gz"</span></span>
<span><span class="co">## [3] "/home/runner/work/_temp/Library/iscream/extdata/cell3.bed.gz"</span></span>
<span><span class="co">## [4] "/home/runner/work/_temp/Library/iscream/extdata/cell4.bed.gz"</span></span></code></pre>
<p>For demonstration, I’m using two regions.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1:184577-680065"</span>, <span class="st">"chrY:56877780-56882524"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Note</strong>: in this dataset the chromosome identifier is
in a “chr[number]” format. However, in some datasets, the chromosome ID
is just a number or a letter like ‘1’, or ‘Y’. To see the chromosome ID
format across your files you can use <code><a href="../reference/query_chroms.html">query_chroms()</a></code> to get
all unique chromosomes across all input files:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/query_chroms.html">query_chroms</a></span><span class="op">(</span><span class="va">bedfiles</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "chr1" "chrY"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="making-queries">Making queries<a class="anchor" aria-label="anchor" href="#making-queries"></a>
</h3>
<p><code><a href="../reference/tabix.html">tabix()</a></code> one file:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tabix.html">tabix</a></span><span class="op">(</span><span class="va">bedfiles</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">regions</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        chr    start      end    V1    V2</span></span>
<span><span class="co">##     &lt;char&gt;    &lt;int&gt;    &lt;int&gt; &lt;num&gt; &lt;int&gt;</span></span>
<span><span class="co">##  1:   chr1   184577   184578     0     1</span></span>
<span><span class="co">##  2:   chr1   191223   191224     0     1</span></span>
<span><span class="co">##  3:   chr1   191264   191265     0     1</span></span>
<span><span class="co">##  4:   chr1   191307   191308     0     1</span></span>
<span><span class="co">##  5:   chr1   191491   191492     0     1</span></span>
<span><span class="co">##  6:   chr1   191507   191508     0     1</span></span>
<span><span class="co">##  7:   chr1   191526   191527     0     1</span></span>
<span><span class="co">##  8:   chr1   679950   679951     1     1</span></span>
<span><span class="co">##  9:   chr1   680064   680065     1     1</span></span>
<span><span class="co">## 10:   chrY 56877780 56877781     0     1</span></span>
<span><span class="co">## 11:   chrY 56878350 56878351     0     1</span></span>
<span><span class="co">## 12:   chrY 56878351 56878352     0     1</span></span>
<span><span class="co">## 13:   chrY 56878431 56878432     0     1</span></span>
<span><span class="co">## 14:   chrY 56879483 56879484     0     2</span></span>
<span><span class="co">## 15:   chrY 56879535 56879536     0     2</span></span>
<span><span class="co">## 16:   chrY 56879539 56879540     1     1</span></span>
<span><span class="co">## 17:   chrY 56879724 56879725     0     1</span></span>
<span><span class="co">## 18:   chrY 56879834 56879835     0     1</span></span>
<span><span class="co">## 19:   chrY 56879863 56879864     1     1</span></span>
<span><span class="co">## 20:   chrY 56879915 56879916     1     1</span></span>
<span><span class="co">## 21:   chrY 56879945 56879946     1     1</span></span>
<span><span class="co">## 22:   chrY 56881130 56881131     1     1</span></span>
<span><span class="co">## 23:   chrY 56881926 56881927     1     1</span></span>
<span><span class="co">## 24:   chrY 56882523 56882524     0     1</span></span>
<span><span class="co">##        chr    start      end    V1    V2</span></span></code></pre>
<p>With multiple files, the output contains a column for the file
name.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tabix.html">tabix</a></span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        chr    start      end    V1    V2   file</span></span>
<span><span class="co">##     &lt;char&gt;    &lt;int&gt;    &lt;int&gt; &lt;num&gt; &lt;int&gt; &lt;char&gt;</span></span>
<span><span class="co">##  1:   chr1   184577   184578     0     1  cell1</span></span>
<span><span class="co">##  2:   chr1   191223   191224     0     1  cell1</span></span>
<span><span class="co">##  3:   chr1   191264   191265     0     1  cell1</span></span>
<span><span class="co">##  4:   chr1   191307   191308     0     1  cell1</span></span>
<span><span class="co">##  5:   chr1   191491   191492     0     1  cell1</span></span>
<span><span class="co">## ---                                            </span></span>
<span><span class="co">## 70:   chrY 56880518 56880519     0     1  cell4</span></span>
<span><span class="co">## 71:   chrY 56880715 56880716     0     2  cell4</span></span>
<span><span class="co">## 72:   chrY 56880914 56880915     0     2  cell4</span></span>
<span><span class="co">## 73:   chrY 56881927 56881928     1     1  cell4</span></span>
<span><span class="co">## 74:   chrY 56882345 56882346     1     1  cell4</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="using-genomicranges">Using <em><a href="https://bioconductor.org/packages/3.21/GenomicRanges" class="external-link">GenomicRanges</a></em><a class="anchor" aria-label="anchor" href="#using-genomicranges"></a>
</h3>
<p>Both <code><a href="../reference/tabix.html">tabix()</a></code> and <code><a href="../reference/tabix.html">tabix_gr()</a></code> accept
<code>GRanges</code> objects as input regions but
<code><a href="../reference/tabix.html">tabix_gr()</a></code> returns <code>GRanges</code> objects instead of
data frames. <code>tabix_gr</code> will also preserve any input
<code>GRanges</code> metadata.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">"GenomicRanges"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"The 'GenomicRanges' package must be installed for this functionality"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="va">regions</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/tabix.html">tabix_gr</a></span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">gr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 4:</span></span>
<span><span class="co">## $cell1</span></span>
<span><span class="co">## GRanges object with 24 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        V1        V2        file</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">##    [1]     chr1    184578      * |         0         1       cell1</span></span>
<span><span class="co">##    [2]     chr1    191224      * |         0         1       cell1</span></span>
<span><span class="co">##    [3]     chr1    191265      * |         0         1       cell1</span></span>
<span><span class="co">##    [4]     chr1    191308      * |         0         1       cell1</span></span>
<span><span class="co">##    [5]     chr1    191492      * |         0         1       cell1</span></span>
<span><span class="co">##    ...      ...       ...    ... .       ...       ...         ...</span></span>
<span><span class="co">##   [20]     chrY  56879916      * |         1         1       cell1</span></span>
<span><span class="co">##   [21]     chrY  56879946      * |         1         1       cell1</span></span>
<span><span class="co">##   [22]     chrY  56881131      * |         1         1       cell1</span></span>
<span><span class="co">##   [23]     chrY  56881927      * |         1         1       cell1</span></span>
<span><span class="co">##   [24]     chrY  56882524      * |         0         1       cell1</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ...</span></span>
<span><span class="co">## &lt;3 more elements&gt;</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="setting-column-names">Setting column names<a class="anchor" aria-label="anchor" href="#setting-column-names"></a>
</h3>
<p>You can set the result data frame’s column names or the
<code>GRanges</code> <code>mcols</code> with <code>col.names</code>:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tabix.html">tabix_gr</a></span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span>, col.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"beta"</span>, <span class="st">"coverage"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 4:</span></span>
<span><span class="co">## $cell1</span></span>
<span><span class="co">## GRanges object with 24 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |      beta  coverage        file</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">##    [1]     chr1    184578      * |         0         1       cell1</span></span>
<span><span class="co">##    [2]     chr1    191224      * |         0         1       cell1</span></span>
<span><span class="co">##    [3]     chr1    191265      * |         0         1       cell1</span></span>
<span><span class="co">##    [4]     chr1    191308      * |         0         1       cell1</span></span>
<span><span class="co">##    [5]     chr1    191492      * |         0         1       cell1</span></span>
<span><span class="co">##    ...      ...       ...    ... .       ...       ...         ...</span></span>
<span><span class="co">##   [20]     chrY  56879916      * |         1         1       cell1</span></span>
<span><span class="co">##   [21]     chrY  56879946      * |         1         1       cell1</span></span>
<span><span class="co">##   [22]     chrY  56881131      * |         1         1       cell1</span></span>
<span><span class="co">##   [23]     chrY  56881927      * |         1         1       cell1</span></span>
<span><span class="co">##   [24]     chrY  56882524      * |         0         1       cell1</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ...</span></span>
<span><span class="co">## &lt;3 more elements&gt;</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="rsamtools-style-output">Rsamtools-style output<a class="anchor" aria-label="anchor" href="#rsamtools-style-output"></a>
</h3>
<p><code><a href="../reference/tabix.html">tabix_raw()</a></code> will return an unparsed named list like
<em><a href="https://bioconductor.org/packages/3.21/Rsamtools" class="external-link">Rsamtools</a></em>,
but with multi-file support:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tabix.html">tabix_raw</a></span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $cell1</span></span>
<span><span class="co">## $cell1$`chr1:184577-680065`</span></span>
<span><span class="co">## [1] "chr1\t184577\t184578\t0.000\t1" "chr1\t191223\t191224\t0.000\t1"</span></span>
<span><span class="co">## [3] "chr1\t191264\t191265\t0.000\t1" "chr1\t191307\t191308\t0.000\t1"</span></span>
<span><span class="co">## [5] "chr1\t191491\t191492\t0.000\t1" "chr1\t191507\t191508\t0.000\t1"</span></span>
<span><span class="co">## [7] "chr1\t191526\t191527\t0.000\t1" "chr1\t679950\t679951\t1.000\t1"</span></span>
<span><span class="co">## [9] "chr1\t680064\t680065\t1.000\t1"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $cell1$`chrY:56877780-56882524`</span></span>
<span><span class="co">##  [1] "chrY\t56877780\t56877781\t0.000\t1" "chrY\t56878350\t56878351\t0.000\t1"</span></span>
<span><span class="co">##  [3] "chrY\t56878351\t56878352\t0.000\t1" "chrY\t56878431\t56878432\t0.000\t1"</span></span>
<span><span class="co">##  [5] "chrY\t56879483\t56879484\t0.000\t2" "chrY\t56879535\t56879536\t0.000\t2"</span></span>
<span><span class="co">##  [7] "chrY\t56879539\t56879540\t1.000\t1" "chrY\t56879724\t56879725\t0.000\t1"</span></span>
<span><span class="co">##  [9] "chrY\t56879834\t56879835\t0.000\t1" "chrY\t56879863\t56879864\t1.000\t1"</span></span>
<span><span class="co">## [11] "chrY\t56879915\t56879916\t1.000\t1" "chrY\t56879945\t56879946\t1.000\t1"</span></span>
<span><span class="co">## [13] "chrY\t56881130\t56881131\t1.000\t1" "chrY\t56881926\t56881927\t1.000\t1"</span></span>
<span><span class="co">## [15] "chrY\t56882523\t56882524\t0.000\t1"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $cell2</span></span>
<span><span class="co">## $cell2$`chr1:184577-680065`</span></span>
<span><span class="co">## character(0)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $cell2$`chrY:56877780-56882524`</span></span>
<span><span class="co">##  [1] "chrY\t56878351\t56878352\t0.000\t1" "chrY\t56879482\t56879483\t1.000\t1"</span></span>
<span><span class="co">##  [3] "chrY\t56879483\t56879484\t1.000\t1" "chrY\t56879534\t56879535\t1.000\t1"</span></span>
<span><span class="co">##  [5] "chrY\t56879538\t56879539\t1.000\t1" "chrY\t56880322\t56880323\t0.000\t1"</span></span>
<span><span class="co">##  [7] "chrY\t56880404\t56880405\t1.000\t1" "chrY\t56881129\t56881130\t0.000\t1"</span></span>
<span><span class="co">##  [9] "chrY\t56881285\t56881286\t0.000\t1" "chrY\t56881364\t56881365\t1.000\t1"</span></span>
<span><span class="co">## [11] "chrY\t56882344\t56882345\t0.000\t1" "chrY\t56882523\t56882524\t0.000\t1"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $cell3</span></span>
<span><span class="co">## $cell3$`chr1:184577-680065`</span></span>
<span><span class="co">## character(0)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $cell3$`chrY:56877780-56882524`</span></span>
<span><span class="co">## [1] "chrY\t56879862\t56879863\t1.000\t1" "chrY\t56879914\t56879915\t0.500\t2"</span></span>
<span><span class="co">## [3] "chrY\t56879944\t56879945\t1.000\t1" "chrY\t56879969\t56879970\t1.000\t1"</span></span>
<span><span class="co">## [5] "chrY\t56879976\t56879977\t1.000\t1" "chrY\t56880322\t56880323\t0.000\t1"</span></span>
<span><span class="co">## [7] "chrY\t56880359\t56880360\t0.000\t1" "chrY\t56881129\t56881130\t0.500\t2"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $cell4</span></span>
<span><span class="co">## $cell4$`chr1:184577-680065`</span></span>
<span><span class="co">##  [1] "chr1\t191003\t191004\t1.000\t1" "chr1\t191307\t191308\t1.000\t1"</span></span>
<span><span class="co">##  [3] "chr1\t191438\t191439\t1.000\t1" "chr1\t191491\t191492\t1.000\t1"</span></span>
<span><span class="co">##  [5] "chr1\t191722\t191723\t1.000\t1" "chr1\t191747\t191748\t1.000\t1"</span></span>
<span><span class="co">##  [7] "chr1\t191776\t191777\t1.000\t1" "chr1\t628458\t628459\t1.000\t1"</span></span>
<span><span class="co">##  [9] "chr1\t631433\t631434\t1.000\t1" "chr1\t631436\t631437\t1.000\t1"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $cell4$`chrY:56877780-56882524`</span></span>
<span><span class="co">##  [1] "chrY\t56877780\t56877781\t0.000\t1" "chrY\t56877810\t56877811\t0.000\t1"</span></span>
<span><span class="co">##  [3] "chrY\t56879834\t56879835\t0.000\t1" "chrY\t56879863\t56879864\t1.000\t1"</span></span>
<span><span class="co">##  [5] "chrY\t56879915\t56879916\t1.000\t1" "chrY\t56879945\t56879946\t0.000\t1"</span></span>
<span><span class="co">##  [7] "chrY\t56880045\t56880046\t0.500\t2" "chrY\t56880081\t56880082\t0.000\t1"</span></span>
<span><span class="co">##  [9] "chrY\t56880092\t56880093\t0.500\t2" "chrY\t56880112\t56880113\t1.000\t1"</span></span>
<span><span class="co">## [11] "chrY\t56880137\t56880138\t1.000\t2" "chrY\t56880142\t56880143\t0.500\t2"</span></span>
<span><span class="co">## [13] "chrY\t56880323\t56880324\t0.000\t1" "chrY\t56880360\t56880361\t1.000\t1"</span></span>
<span><span class="co">## [15] "chrY\t56880405\t56880406\t1.000\t1" "chrY\t56880518\t56880519\t0.000\t1"</span></span>
<span><span class="co">## [17] "chrY\t56880715\t56880716\t0.000\t2" "chrY\t56880914\t56880915\t0.000\t2"</span></span>
<span><span class="co">## [19] "chrY\t56881927\t56881928\t1.000\t1" "chrY\t56882345\t56882346\t1.000\t1"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="wgbs-bed-files">WGBS BED files<a class="anchor" aria-label="anchor" href="#wgbs-bed-files"></a>
</h3>
<p>Since iscream was originally developed to read Whole Genome Bisulfite
Sequencing data, it has the <code>aligner</code> argument for
automatically setting column names for the BISCUIT <span class="citation">(<a href="#ref-zhou2024">Zhou et al. 2024</a>)</span>,
Bismark <span class="citation">(<a href="#ref-krueger2011">Krueger and
Andrews 2011</a>)</span>, and BSBolt <span class="citation">(<a href="#ref-farrell2021">Farrell et al. 2021</a>)</span>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tabix.html">tabix_gr</a></span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span>, aligner <span class="op">=</span> <span class="st">"biscuit"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 4:</span></span>
<span><span class="co">## $cell1</span></span>
<span><span class="co">## GRanges object with 24 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |      beta  coverage        file</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">##    [1]     chr1    184578      * |         0         1       cell1</span></span>
<span><span class="co">##    [2]     chr1    191224      * |         0         1       cell1</span></span>
<span><span class="co">##    [3]     chr1    191265      * |         0         1       cell1</span></span>
<span><span class="co">##    [4]     chr1    191308      * |         0         1       cell1</span></span>
<span><span class="co">##    [5]     chr1    191492      * |         0         1       cell1</span></span>
<span><span class="co">##    ...      ...       ...    ... .       ...       ...         ...</span></span>
<span><span class="co">##   [20]     chrY  56879916      * |         1         1       cell1</span></span>
<span><span class="co">##   [21]     chrY  56879946      * |         1         1       cell1</span></span>
<span><span class="co">##   [22]     chrY  56881131      * |         1         1       cell1</span></span>
<span><span class="co">##   [23]     chrY  56881927      * |         1         1       cell1</span></span>
<span><span class="co">##   [24]     chrY  56882524      * |         0         1       cell1</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ...</span></span>
<span><span class="co">## &lt;3 more elements&gt;</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="summarize_regions">
<code>summarize_regions()</code><a class="anchor" aria-label="anchor" href="#summarize_regions"></a>
</h2>
<p>Using the same BED files and regions, iscream can also summarize the
data within regions. This requires providing the index of the data
columns you want summarized and the column names for the output. All
summarizing functions are run on each input column by default - see
<code><a href="../reference/summarize_regions.html">?summarize_regions</a></code> for supported functions.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summarize_regions.html">summarize_regions</a></span><span class="op">(</span></span>
<span>  <span class="va">bedfiles</span>,</span>
<span>  <span class="va">regions</span>,</span>
<span>  columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">5</span><span class="op">)</span>,</span>
<span>  col_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"beta"</span>, <span class="st">"coverage"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [19:48:10.559316] [iscream::summarize_regions] [info] Summarizing 2 regions from 4 bedfiles</span></span>
<span><span class="co">## [19:48:10.559376] [iscream::summarize_regions] [info] using sum, mean, median, stddev, variance, min, max, range, count</span></span>
<span><span class="co">## [19:48:10.559380] [iscream::summarize_regions] [info] with columns 4, 5 as beta, coverage</span></span></code></pre>
<pre><code><span><span class="co">##                  feature  file beta.sum coverage.sum beta.mean coverage.mean</span></span>
<span><span class="co">## 1     chr1:184577-680065 cell1      2.0            9 0.2222222      1.000000</span></span>
<span><span class="co">## 2 chrY:56877780-56882524 cell1      6.0           17 0.4000000      1.133333</span></span>
<span><span class="co">## 3     chr1:184577-680065 cell2       NA           NA        NA            NA</span></span>
<span><span class="co">## 4 chrY:56877780-56882524 cell2      6.0           12 0.5000000      1.000000</span></span>
<span><span class="co">## 5     chr1:184577-680065 cell3       NA           NA        NA            NA</span></span>
<span><span class="co">## 6 chrY:56877780-56882524 cell3      5.0           10 0.6250000      1.250000</span></span>
<span><span class="co">## 7     chr1:184577-680065 cell4     10.0           10 1.0000000      1.000000</span></span>
<span><span class="co">## 8 chrY:56877780-56882524 cell4      9.5           26 0.4750000      1.300000</span></span>
<span><span class="co">##   beta.median coverage.median beta.stddev coverage.stddev beta.variance</span></span>
<span><span class="co">## 1        0.00               1   0.4409586       0.0000000     0.1944444</span></span>
<span><span class="co">## 2        0.00               1   0.5070926       0.3518658     0.2571429</span></span>
<span><span class="co">## 3          NA              NA          NA              NA            NA</span></span>
<span><span class="co">## 4        0.50               1   0.5222330       0.0000000     0.2727273</span></span>
<span><span class="co">## 5          NA              NA          NA              NA            NA</span></span>
<span><span class="co">## 6        0.75               1   0.4432026       0.4629100     0.1964286</span></span>
<span><span class="co">## 7        1.00               1   0.0000000       0.0000000     0.0000000</span></span>
<span><span class="co">## 8        0.50               1   0.4722566       0.4701623     0.2230263</span></span>
<span><span class="co">##   coverage.variance beta.min coverage.min beta.max coverage.max beta.range</span></span>
<span><span class="co">## 1         0.0000000        0            1        1            1          1</span></span>
<span><span class="co">## 2         0.1238095        0            1        1            2          1</span></span>
<span><span class="co">## 3                NA       NA           NA       NA           NA         NA</span></span>
<span><span class="co">## 4         0.0000000        0            1        1            1          1</span></span>
<span><span class="co">## 5                NA       NA           NA       NA           NA         NA</span></span>
<span><span class="co">## 6         0.2142857        0            1        1            2          1</span></span>
<span><span class="co">## 7         0.0000000        1            1        1            1          0</span></span>
<span><span class="co">## 8         0.2210526        0            1        1            2          1</span></span>
<span><span class="co">##   coverage.range count</span></span>
<span><span class="co">## 1              0     9</span></span>
<span><span class="co">## 2              1    15</span></span>
<span><span class="co">## 3             NA    NA</span></span>
<span><span class="co">## 4              0    12</span></span>
<span><span class="co">## 5             NA    NA</span></span>
<span><span class="co">## 6              1     8</span></span>
<span><span class="co">## 7              0    10</span></span>
<span><span class="co">## 8              1    20</span></span></code></pre>
<p>The <code>feature</code> column here contains the genomic region
coordinates, but can be set to something more informational if you have
names for the regions. You can also select the functions applied with
<code>fun</code>:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">regions</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"R1"</span>, <span class="st">"R2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/summarize_regions.html">summarize_regions</a></span><span class="op">(</span></span>
<span>  <span class="va">bedfiles</span>,</span>
<span>  <span class="va">regions</span>,</span>
<span>  fun <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mean"</span>, <span class="st">"sum"</span><span class="op">)</span>,</span>
<span>  columns <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  col_names <span class="op">=</span> <span class="st">"coverage"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [19:48:10.649416] [iscream::summarize_regions] [info] Summarizing 2 regions from 4 bedfiles</span></span>
<span><span class="co">## [19:48:10.649437] [iscream::summarize_regions] [info] using mean, sum</span></span>
<span><span class="co">## [19:48:10.649441] [iscream::summarize_regions] [info] with columns 5 as coverage</span></span></code></pre>
<pre><code><span><span class="co">##   feature  file coverage.mean coverage.sum</span></span>
<span><span class="co">## 1      R1 cell1      1.000000            9</span></span>
<span><span class="co">## 2      R2 cell1      1.133333           17</span></span>
<span><span class="co">## 3      R1 cell2            NA           NA</span></span>
<span><span class="co">## 4      R2 cell2      1.000000           12</span></span>
<span><span class="co">## 5      R1 cell3            NA           NA</span></span>
<span><span class="co">## 6      R2 cell3      1.250000           10</span></span>
<span><span class="co">## 7      R1 cell4      1.000000           10</span></span>
<span><span class="co">## 8      R2 cell4      1.300000           26</span></span></code></pre>
<div class="section level3">
<h3 id="wgbs-bed-files-1">WGBS BED files<a class="anchor" aria-label="anchor" href="#wgbs-bed-files-1"></a>
</h3>
<p>For WGBS data specifically, you can use
<code><a href="../reference/summarize_meth_regions.html">summarize_meth_regions()</a></code>, which takes the
<code>aligner</code> argument to correctly parse the columns.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summarize_meth_regions.html">summarize_meth_regions</a></span><span class="op">(</span></span>
<span>  <span class="va">bedfiles</span>,</span>
<span>  <span class="va">regions</span>,</span>
<span>  aligner <span class="op">=</span> <span class="st">"biscuit"</span>,</span>
<span>  fun <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mean"</span>, <span class="st">"sum"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [19:48:10.718610] [iscream::summarize_regions] [info] Summarizing 2 regions from 4 bedfiles</span></span>
<span><span class="co">## [19:48:10.718631] [iscream::summarize_regions] [info] using mean, sum</span></span>
<span><span class="co">## [19:48:10.718634] [iscream::summarize_regions] [info] with columns 4, 5 as coverage, M</span></span></code></pre>
<pre><code><span><span class="co">##   feature  file coverage.mean    M.mean coverage.sum M.sum</span></span>
<span><span class="co">## 1      R1 cell1      1.000000 0.2222222            9     2</span></span>
<span><span class="co">## 2      R2 cell1      1.133333 0.4000000           17     6</span></span>
<span><span class="co">## 3      R1 cell2            NA        NA           NA    NA</span></span>
<span><span class="co">## 4      R2 cell2      1.000000 0.5000000           12     6</span></span>
<span><span class="co">## 5      R1 cell3            NA        NA           NA    NA</span></span>
<span><span class="co">## 6      R2 cell3      1.250000 0.7500000           10     6</span></span>
<span><span class="co">## 7      R1 cell4      1.000000 1.0000000           10    10</span></span>
<span><span class="co">## 8      R2 cell4      1.300000 0.6000000           26    12</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="make_mat">
<code>make_mat()</code><a class="anchor" aria-label="anchor" href="#make_mat"></a>
</h2>
<p><code><a href="../reference/make_mat.html">make_mat()</a></code> makes a matrix where every row is a locus
found in at least one input file and the columns are the input files. It
returns a named list of the matrix, coordinates, and file names you can
use for analysis or other data structures. <code><a href="../reference/make_mat.html">make_mat_se()</a></code>
returns a <code>RangedSummarizedExperiment</code>, and
<code><a href="../reference/make_mat.html">make_mat_gr()</a></code> returns a <code>GRanges</code> object.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">"SummarizedExperiment"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"The 'SummarizedExperiment' package must be installed for this functionality"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="op">(</span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_mat.html">make_mat_se</a></span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span>, column <span class="op">=</span> <span class="fl">4</span>, mat_name <span class="op">=</span> <span class="st">"beta"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [19:48:12.963113] [iscream::query_all] [info] Querying 2 regions from 4 bedfiles</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [19:48:12.963639] [iscream::query_all] [info] Creating metadata vectors</span></span>
<span><span class="co">## [19:48:12.963698] [iscream::query_all] [info] 62 loci found - 9938 extra rows allocated with 0 resizes</span></span>
<span><span class="co">## [19:48:12.963703] [iscream::query_all] [info] Creating dense matrix</span></span></code></pre>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 62 4 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): beta</span></span>
<span><span class="co">## rownames: NULL</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(4): cell1 cell2 cell3 cell4</span></span>
<span><span class="co">## colData names(0):</span></span></code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       cell1 cell2 cell3 cell4</span></span>
<span><span class="co">##  [1,]     0     0   1.0     0</span></span>
<span><span class="co">##  [2,]     0     0   0.5     0</span></span>
<span><span class="co">##  [3,]     0     0   1.0     0</span></span>
<span><span class="co">##  [4,]     0     0   1.0     0</span></span>
<span><span class="co">##  [5,]     0     0   1.0     0</span></span>
<span><span class="co">##  [6,]     0     0   0.0     0</span></span>
<span><span class="co">##  [7,]     0     0   0.0     0</span></span>
<span><span class="co">##  [8,]     0     0   0.5     0</span></span>
<span><span class="co">##  [9,]     0     0   0.0     0</span></span>
<span><span class="co">## [10,]     0     0   0.0     0</span></span></code></pre>
<p>If you have sparse data, you can save memory with
<code>sparse = TRUE</code>, but only for <code><a href="../reference/make_mat.html">make_mat()</a></code> and
<code><a href="../reference/make_mat.html">make_mat_se()</a></code>.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_mat.html">make_mat</a></span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span>, column <span class="op">=</span> <span class="fl">4</span>, mat_name <span class="op">=</span> <span class="st">"beta"</span>, sparse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [19:48:13.081637] [iscream::query_all] [info] Querying 2 regions from 4 bedfiles</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [19:48:13.082110] [iscream::query_all] [info] Creating metadata vectors</span></span>
<span><span class="co">## [19:48:13.082145] [iscream::query_all] [info] 62 loci found - 9938 extra rows allocated with 0 resizes</span></span>
<span><span class="co">## [19:48:13.082156] [iscream::query_all] [info] Creating sparse matrix</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mat</span><span class="op">$</span><span class="va">beta</span>, <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 10 x 4 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##       cell1 cell2 cell3 cell4</span></span>
<span><span class="co">##  [1,]     .     .   1.0     .</span></span>
<span><span class="co">##  [2,]     .     .   0.5     .</span></span>
<span><span class="co">##  [3,]     .     .   1.0     .</span></span>
<span><span class="co">##  [4,]     .     .   1.0     .</span></span>
<span><span class="co">##  [5,]     .     .   1.0     .</span></span>
<span><span class="co">##  [6,]     .     .   .       .</span></span>
<span><span class="co">##  [7,]     .     .   .       .</span></span>
<span><span class="co">##  [8,]     .     .   0.5     .</span></span>
<span><span class="co">##  [9,]     .     .   .       .</span></span>
<span><span class="co">## [10,]     .     .   .       .</span></span></code></pre>
<div class="section level3">
<h3 id="wgbs-bed-files-2">WGBS BED files<a class="anchor" aria-label="anchor" href="#wgbs-bed-files-2"></a>
</h3>
<p>For a <em><a href="https://bioconductor.org/packages/3.21/BSseq" class="external-link">BSseq</a></em>
compatible object, use <code>make_meth_mat</code> which returns a list
of BSseq inputs. To make the <code>BSseq</code> object run</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/kasperdanielhansen/bsseq" class="external-link">"bsseq"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">meth_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_mat_bsseq.html">make_mat_bsseq</a></span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">BSseq</span>, <span class="va">meth_mat</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<pre><code><span><span class="co">## [19:48:16.928145] [iscream::query_all] [info] Querying 2 regions from 4 bedfiles</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [19:48:16.928675] [iscream::query_all] [info] Creating metadata vectors</span></span>
<span><span class="co">## [19:48:16.928713] [iscream::query_all] [info] 62 loci found - 9938 extra rows allocated with 0 resizes</span></span>
<span><span class="co">## [19:48:16.928718] [iscream::query_all] [info] Creating dense matrix</span></span></code></pre>
<pre><code><span><span class="co">## An object of type 'BSseq' with</span></span>
<span><span class="co">##   62 methylation loci</span></span>
<span><span class="co">##   4 samples</span></span>
<span><span class="co">## has not been smoothed</span></span>
<span><span class="co">## All assays are in-memory</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="further-reading">Further reading<a class="anchor" aria-label="anchor" href="#further-reading"></a>
</h2>
<p>While this vignette used a toy dataset,
<code><a href="../articles/performance.html">vignette("performance")</a></code> has tips on getting the best
performance from iscream using a real and larger dataset with 100 files.
<code><a href="../articles/TSS.html">vignette("TSS")</a></code> is an example of iscream’s application to
examine methylation around transcription start site on another real
dataset. For more information on the supported Bioconductor data
structures and conversions see
<code><a href="../articles/data_structures.html">vignette("data_structures")</a></code>.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] bsseq_1.44.1                SummarizedExperiment_1.38.1</span></span>
<span><span class="co">##  [3] Biobase_2.68.0              MatrixGenerics_1.20.0      </span></span>
<span><span class="co">##  [5] matrixStats_1.5.0           GenomicRanges_1.60.0       </span></span>
<span><span class="co">##  [7] GenomeInfoDb_1.44.3         IRanges_2.42.0             </span></span>
<span><span class="co">##  [9] S4Vectors_0.46.0            BiocGenerics_0.54.1        </span></span>
<span><span class="co">## [11] generics_0.1.4              iscream_0.99.10            </span></span>
<span><span class="co">## [13] BiocStyle_2.36.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] farver_2.1.2              R.utils_2.13.0           </span></span>
<span><span class="co">##  [3] Biostrings_2.76.0         bitops_1.0-9             </span></span>
<span><span class="co">##  [5] fastmap_1.2.0             RCurl_1.98-1.17          </span></span>
<span><span class="co">##  [7] GenomicAlignments_1.44.0  stringfish_0.17.0        </span></span>
<span><span class="co">##  [9] XML_3.99-0.19             digest_0.6.37            </span></span>
<span><span class="co">## [11] lifecycle_1.0.4           statmod_1.5.1            </span></span>
<span><span class="co">## [13] compiler_4.5.1            rlang_1.1.6              </span></span>
<span><span class="co">## [15] sass_0.4.10               tools_4.5.1              </span></span>
<span><span class="co">## [17] yaml_2.3.10               data.table_1.17.8        </span></span>
<span><span class="co">## [19] rtracklayer_1.68.0        knitr_1.50               </span></span>
<span><span class="co">## [21] S4Arrays_1.8.1            curl_7.0.0               </span></span>
<span><span class="co">## [23] DelayedArray_0.34.1       RColorBrewer_1.1-3       </span></span>
<span><span class="co">## [25] abind_1.4-8               BiocParallel_1.42.2      </span></span>
<span><span class="co">## [27] HDF5Array_1.36.0          R.oo_1.27.1              </span></span>
<span><span class="co">## [29] desc_1.4.3                grid_4.5.1               </span></span>
<span><span class="co">## [31] beachmat_2.24.0           Rhdf5lib_1.30.0          </span></span>
<span><span class="co">## [33] scales_1.4.0              gtools_3.9.5             </span></span>
<span><span class="co">## [35] cli_3.6.5                 rmarkdown_2.30           </span></span>
<span><span class="co">## [37] crayon_1.5.3              ragg_1.5.0               </span></span>
<span><span class="co">## [39] RcppParallel_5.1.11-1     rjson_0.2.23             </span></span>
<span><span class="co">## [41] httr_1.4.7                DelayedMatrixStats_1.30.0</span></span>
<span><span class="co">## [43] pbapply_1.7-4             cachem_1.1.0             </span></span>
<span><span class="co">## [45] rhdf5_2.52.1              parallel_4.5.1           </span></span>
<span><span class="co">## [47] BiocManager_1.30.26       XVector_0.48.0           </span></span>
<span><span class="co">## [49] restfulr_0.0.16           Matrix_1.7-3             </span></span>
<span><span class="co">## [51] jsonlite_2.0.0            bookdown_0.45            </span></span>
<span><span class="co">## [53] systemfonts_1.3.1         h5mread_1.0.1            </span></span>
<span><span class="co">## [55] locfit_1.5-9.12           limma_3.64.3             </span></span>
<span><span class="co">## [57] jquerylib_0.1.4           glue_1.8.0               </span></span>
<span><span class="co">## [59] parallelly_1.45.1         pkgdown_2.1.3            </span></span>
<span><span class="co">## [61] codetools_0.2-20          BiocIO_1.18.0            </span></span>
<span><span class="co">## [63] UCSC.utils_1.4.0          htmltools_0.5.8.1        </span></span>
<span><span class="co">## [65] rhdf5filters_1.20.0       GenomeInfoDbData_1.2.14  </span></span>
<span><span class="co">## [67] BSgenome_1.76.0           R6_2.6.1                 </span></span>
<span><span class="co">## [69] textshaping_1.0.4         sparseMatrixStats_1.20.0 </span></span>
<span><span class="co">## [71] evaluate_1.0.5            lattice_0.22-7           </span></span>
<span><span class="co">## [73] R.methodsS3_1.8.2         Rsamtools_2.24.1         </span></span>
<span><span class="co">## [75] bslib_0.9.0               Rcpp_1.1.0               </span></span>
<span><span class="co">## [77] SparseArray_1.8.1         permute_0.9-8            </span></span>
<span><span class="co">## [79] xfun_0.53                 fs_1.6.6</span></span></code></pre>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-farrell2021" class="csl-entry">
Farrell, Colin, Michael Thompson, Anela Tosevska, Adewale Oyetunde, and
Matteo Pellegrini. 2021. <span>“<span>BiSulfite Bolt</span>:
<span>A</span> Bisulfite Sequencing Analysis Platform.”</span>
<em>Gigascience</em> 10 (5): giab033. <a href="https://doi.org/10.1093/gigascience/giab033" class="external-link">https://doi.org/10.1093/gigascience/giab033</a>.
</div>
<div id="ref-krueger2011" class="csl-entry">
Krueger, Felix, and Simon R. Andrews. 2011. <span>“Bismark: A Flexible
Aligner and Methylation Caller for <span>Bisulfite-Seq</span>
Applications.”</span> <em>Bioinformatics</em> 27 (11): 1571–72. <a href="https://doi.org/10.1093/bioinformatics/btr167" class="external-link">https://doi.org/10.1093/bioinformatics/btr167</a>.
</div>
<div id="ref-luo2018a" class="csl-entry">
Luo, Chongyuan, Angeline Rivkin, Jingtian Zhou, Justin P. Sandoval,
Laurie Kurihara, Jacinta Lucero, Rosa Castanon, et al. 2018.
<span>“Robust Single-Cell <span>DNA</span> Methylome Profiling with
<span class="nocase">snmC-seq2</span>.”</span> <em>Nat Commun</em> 9
(1): 3824. <a href="https://doi.org/10.1038/s41467-018-06355-2" class="external-link">https://doi.org/10.1038/s41467-018-06355-2</a>.
</div>
<div id="ref-zhou2024" class="csl-entry">
Zhou, Wanding, Benjamin K Johnson, Jacob Morrison, Ian Beddows, James
Eapen, Efrat Katsman, Ayush Semwal, et al. 2024.
<span>“<span>BISCUIT</span>: An Efficient, Standards-Compliant Tool
Suite for Simultaneous Genetic and Epigenetic Inference in Bulk and
Single-Cell Studies.”</span> <em>Nucleic Acids Research</em> 52 (6):
gkae097. <a href="https://doi.org/10.1093/nar/gkae097" class="external-link">https://doi.org/10.1093/nar/gkae097</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by James Eapen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
