<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Getting started with iscream • iscream</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="../deps/MathJax-3.2.2/tex-chtml.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting started with iscream">
<meta name="robots" content="noindex">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">iscream</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/iscream.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/data_structures.html">Getting other data structures from iscream</a></li>
    <li><a class="dropdown-item" href="../articles/TSS.html">Plotting TSS methylation profiles</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/huishenlab/iscream/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Getting started with iscream</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/huishenlab/iscream/blob/dev/vignettes/iscream.Rmd" class="external-link"><code>vignettes/iscream.Rmd</code></a></small>
      <div class="d-none name"><code>iscream.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<div class="section level3">
<h3 id="loading-iscream">Loading <em>iscream</em><a class="anchor" aria-label="anchor" href="#loading-iscream"></a>
</h3>
<p>The number of threads can be set before or after loading the
library:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span><span class="st">"iscream.threads"</span> <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://huishenlab.github.io/iscream/">iscream</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; iscream using 8 threads based on 'options(iscream.threads)' but parallelly::availableCores() detects 16 possibly available threads. See `?set_threads` for information on multithreading before trying to use more.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="input-bedifiles">Input bedifiles<a class="anchor" aria-label="anchor" href="#input-bedifiles"></a>
</h3>
<p>I’m using 100 human cell data from the snmc-seq2 dataset in this
vignette:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bedfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span></span>
<span>    <span class="st">"../data/k562/"</span>,</span>
<span>    pattern <span class="op">=</span> <span class="st">"*_mergecg.bed.gz$"</span>,</span>
<span>    full.names <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="regions">Regions<a class="anchor" aria-label="anchor" href="#regions"></a>
</h3>
<p>Since <em>iscream</em> is a region-based querying tool, we need to
load some regions. Here we’ll be using gene body regions as the
input:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/data.table/man/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"../data/genes.bed"</span>, col.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"gene"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">regions</span><span class="op">)</span></span>
<span><span class="co">#&gt;       chr    start      end   gene</span></span>
<span><span class="co">#&gt;    &lt;char&gt;    &lt;int&gt;    &lt;int&gt; &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:   chr1  1471764  1497848 ATAD3B</span></span>
<span><span class="co">#&gt; 2:   chr1  3069167  3438621 PRDM16</span></span>
<span><span class="co">#&gt; 3:   chr1  2403963  2413797  PEX10</span></span>
<span><span class="co">#&gt; 4:   chr1 10472287 10630758  PEX14</span></span>
<span><span class="co">#&gt; 5:   chr1  2425979  2505532  PLCH2</span></span>
<span><span class="co">#&gt; 6:   chr1  9292893  9369532  SPSB1</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="running-iscream">Running iscream<a class="anchor" aria-label="anchor" href="#running-iscream"></a>
</h2>
<div class="section level3">
<h3 id="make-tabix-queries">Make tabix queries<a class="anchor" aria-label="anchor" href="#make-tabix-queries"></a>
</h3>
<p>The <code><a href="../reference/tabix.html">tabix()</a></code> function can be used to query regions from
BED files much like the <em>tabix</em> shell command. It returns the
queried lines from the BED files, parsed into columns, as a
<code>data.table</code>. <code><a href="../reference/tabix.html">tabix()</a></code> is a generic BED files
query function and is not restricted to methylation BED files. It has
built-in support for BISCUIT, Bismark, and BSbolt BED file column names,
set with the <code>aligner</code> argument, but can take other column
names with the <code>colnames</code> argument.</p>
<p>If multiple input files are provided, they are queried in parallel.
If <code>raw = TRUE</code>, <code><a href="../reference/tabix.html">tabix()</a></code> will return the same
data as <code><a href="https://rdrr.io/pkg/Rsamtools/man/scanTabix.html" class="external-link">Rsamtools::scanTabix()</a></code> does - a named list of
strings.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tabix.html">tabix</a></span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;          chr    start      end  beta coverage         mergecg      sample</span></span>
<span><span class="co">#&gt;       &lt;char&gt;    &lt;num&gt;    &lt;num&gt; &lt;num&gt;    &lt;num&gt;          &lt;char&gt;      &lt;char&gt;</span></span>
<span><span class="co">#&gt;    1:   chr1  3147020  3147022 0.000        3 C:0.000:3,G:.:0 A10_mergecg</span></span>
<span><span class="co">#&gt;    2:   chr1  3147050  3147052 0.333        3 C:0.333:3,G:.:0 A10_mergecg</span></span>
<span><span class="co">#&gt;    3:   chr1  3147084  3147086 0.250        4 C:0.250:4,G:.:0 A10_mergecg</span></span>
<span><span class="co">#&gt;    4:   chr1  3147088  3147090 0.250        4 C:0.250:4,G:.:0 A10_mergecg</span></span>
<span><span class="co">#&gt;    5:   chr1  3147106  3147108 0.750        4 C:0.750:4,G:.:0 A10_mergecg</span></span>
<span><span class="co">#&gt;   ---                                                                    </span></span>
<span><span class="co">#&gt; 4276:   chr1 30875221 30875223 1.000        4 C:.:0,G:1.000:4  H9_mergecg</span></span>
<span><span class="co">#&gt; 4277:   chr1 30875294 30875296 0.000        4 C:.:0,G:0.000:4  H9_mergecg</span></span>
<span><span class="co">#&gt; 4278:   chr1 30875303 30875305 0.000        3 C:.:0,G:0.000:3  H9_mergecg</span></span>
<span><span class="co">#&gt; 4279:   chr1  9558942  9558944 1.000        5 C:.:0,G:1.000:5  H9_mergecg</span></span>
<span><span class="co">#&gt; 4280:   chr1 15556745 15556747 0.000        4 C:0.000:4,G:.:0  H9_mergecg</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="get-summary-data">Get summary data<a class="anchor" aria-label="anchor" href="#get-summary-data"></a>
</h3>
<p>To get a summary of the methylation information of the gene bodies we
use <code>summarize_regions</code>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">full_summary</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_regions.html">summarize_regions</a></span><span class="op">(</span><span class="va">bedfiles</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span>, <span class="va">regions</span><span class="op">)</span></span>
<span><span class="co">#&gt; [10:10:12.814740] [iscream::summarize_regions] [info] Summarizing 20048 regions from 8 bedfiles</span></span>
<span><span class="co">#&gt; [10:10:12.814782] [iscream::summarize_regions] [info] using sum, mean, median, stddev, variance, min, max, range, cpg_count</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="build-matrices">Build matrices<a class="anchor" aria-label="anchor" href="#build-matrices"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meth_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/query_all.html">query_all</a></span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span>, sparse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [10:10:14.084742] [iscream::query_all] [info] Querying 20048 regions from 94 bedfiles</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [10:10:21.241229] [iscream::query_all] [info] Creating metadata vectors</span></span>
<span><span class="co">#&gt; [10:10:21.263587] [iscream::query_all] [info] nrows 800000 - 32332 extra rows allocated with 26 resizes</span></span>
<span><span class="co">#&gt; [10:10:21.337022] [iscream::query_all] [info] Creating sparse matrix</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">meth_mat</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 5</span></span>
<span><span class="co">#&gt;  $ M          :Formal class 'dgCMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">#&gt;   .. ..@ i       : int [1:354881] 11 16 1204 4595 4596 4752 5522 5523 6105 6106 ...</span></span>
<span><span class="co">#&gt;   .. ..@ p       : int [1:95] 0 12632 14578 14580 23398 28702 33905 38840 48626 49353 ...</span></span>
<span><span class="co">#&gt;   .. ..@ Dim     : int [1:2] 767668 94</span></span>
<span><span class="co">#&gt;   .. ..@ Dimnames:List of 2</span></span>
<span><span class="co">#&gt;   .. .. ..$ : NULL</span></span>
<span><span class="co">#&gt;   .. .. ..$ : chr [1:94] "A10_mergecg" "A11_mergecg" "A12_mergecg" "A2_mergecg" ...</span></span>
<span><span class="co">#&gt;   .. ..@ x       : num [1:354881] 10 326 3 2 3 4 3 3 2 3 ...</span></span>
<span><span class="co">#&gt;   .. ..@ factors : list()</span></span>
<span><span class="co">#&gt;  $ Cov        :Formal class 'dgCMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">#&gt;   .. ..@ i       : int [1:809902] 10 11 15 16 17 1204 1369 1371 1500 3032 ...</span></span>
<span><span class="co">#&gt;   .. ..@ p       : int [1:95] 0 30350 33083 33088 62181 69341 83514 98102 113013 113984 ...</span></span>
<span><span class="co">#&gt;   .. ..@ Dim     : int [1:2] 767668 94</span></span>
<span><span class="co">#&gt;   .. ..@ Dimnames:List of 2</span></span>
<span><span class="co">#&gt;   .. .. ..$ : NULL</span></span>
<span><span class="co">#&gt;   .. .. ..$ : chr [1:94] "A10_mergecg" "A11_mergecg" "A12_mergecg" "A2_mergecg" ...</span></span>
<span><span class="co">#&gt;   .. ..@ x       : num [1:809902] 56 10 202 326 345 3 3 3 3 3 ...</span></span>
<span><span class="co">#&gt;   .. ..@ factors : list()</span></span>
<span><span class="co">#&gt;  $ pos        : int [1:767668] 162072285 143360905 143360940 54939858 182194770 182194778 182194861 182194913 23391102 123741210 ...</span></span>
<span><span class="co">#&gt;  $ chr        : chr [1:767668] "chr1" "chr3" "chr3" "chr3" ...</span></span>
<span><span class="co">#&gt;  $ sampleNames: chr [1:94] "A10_mergecg" "A11_mergecg" "A12_mergecg" "A2_mergecg" ...</span></span></code></pre></div>
<p>The output of query_all is a named list containing matrices of
coverage values and M values and vectors of the sample names, chromosome
names and positions of the loci.</p>
<p>This list can be used to produce a BSseq object. However since BSseq
cannot work with sparse matrices, the two matrices would need to be
converted to dense matrices first.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">BSseq</span>, <span class="va">meth_mat</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.1 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS/LAPACK: /nix/store/p2a6x92n2gwljav85zz3hzid1nk7iq0l-blas-3/lib/libblas.so.3;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/Detroit</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] data.table_1.15.4  iscream_0.0.0.9000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] compiler_4.4.1     parallelly_1.38.0  Matrix_1.7-0       parallel_4.4.1    </span></span>
<span><span class="co">#&gt;  [5] tools_4.4.1        Rcpp_1.0.13        grid_4.4.1         knitr_1.47        </span></span>
<span><span class="co">#&gt;  [9] xfun_0.45          RcppParallel_5.1.8 stringfish_0.16.0  lattice_0.22-6    </span></span>
<span><span class="co">#&gt; [13] evaluate_0.24.0</span></span></code></pre></div>
<!-- vim: set filetype=rmd: -->
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by James Eapen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
