<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Manuscript data availabiliy • iscream</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Manuscript data availabiliy">
<meta name="robots" content="noindex">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">iscream</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">0.99.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/iscream.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/performance.html">Improving iscream performance</a></li>
    <li><a class="dropdown-item" href="../articles/data_structures.html">iscream compatible data structures</a></li>
    <li><a class="dropdown-item" href="../articles/TSS.html">Plotting TSS methylation profiles</a></li>
    <li><a class="dropdown-item" href="../articles/tabix.html">iscream vs Rsamtools::scanTabix</a></li>
    <li><a class="dropdown-item" href="../articles/htslib.html">Getting htslib headers</a></li>
    <li><a class="dropdown-item" href="../articles/manuscript_data.html">Manuscript data availabiliy</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/huishenlab/iscream/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Manuscript data availabiliy</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/huishenlab/iscream/blob/dev/vignettes/manuscript_data.Rmd" class="external-link"><code>vignettes/manuscript_data.Rmd</code></a></small>
      <div class="d-none name"><code>manuscript_data.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="datasets">Datasets<a class="anchor" aria-label="anchor" href="#datasets"></a>
</h2>
<p>Two datasets were used to produce the figure in the paper:</p>
<ol style="list-style-type: decimal">
<li>snmC-seq2 scWGBS data <span class="citation">(<a href="#ref-luo2018a">Luo et al. 2018</a>)</span> for single-cell
benchmarks.</li>
<li>bulk WGBS from a project in our lab</li>
</ol>
<p>Both datasets were aligned with BISCUIT <span class="citation">(<a href="#ref-zhou2024">Zhou et al. 2024</a>)</span>. To produce Bismark
BEDgraph files <span class="citation">(<a href="#ref-krueger2011">Krueger and Andrews 2011</a>)</span>, we used a
python script to convert the beta and coverage value columns to
percentage methylation, unmethylated reads and methylated read
columns.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode py"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="im">import</span> sys</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="im">import</span> gzip</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>bedfile <span class="op">=</span> sys.argv[<span class="dv">1</span>]</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="kw">def</span> bed2cov(line):</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>    <span class="bu">chr</span>, start, end, beta, cov <span class="op">=</span> line.split(<span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>    percent, meth, unmeth <span class="op">=</span> convert(<span class="bu">float</span>(beta), <span class="bu">int</span>(cov))</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>    <span class="cf">return</span> <span class="st">'</span><span class="ch">\t</span><span class="st">'</span>.join([<span class="bu">chr</span>, start, start, <span class="bu">str</span>(percent), <span class="bu">str</span>(meth), <span class="bu">str</span>(unmeth)])</span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="kw">def</span> convert(beta, cov):</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a>    percent <span class="op">=</span> <span class="bu">round</span>(beta <span class="op">*</span> <span class="dv">100</span>)</span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a>    meth <span class="op">=</span> <span class="bu">round</span>(beta <span class="op">*</span> cov)</span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a>    unmeth <span class="op">=</span> cov <span class="op">-</span> meth</span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a>    <span class="cf">return</span> [percent, meth, unmeth]</span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a><span class="cf">with</span> gzip.<span class="bu">open</span>(bedfile, <span class="st">'rt'</span>) <span class="im">as</span> bed:</span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a>    <span class="cf">for</span> line <span class="kw">in</span> bed:</span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a>        <span class="bu">print</span>(bed2cov(line))</span></code></pre></div>
<p>Using GNU parallel:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">parallel</span> python bed2cov.py <span class="op">&lt;</span>biscuit_path<span class="op">&gt;</span>/{} <span class="st">'&gt;'</span> <span class="op">&lt;</span>bismark_path<span class="op">&gt;</span>/{/.} ::: biscuit_path/<span class="pp">*</span>.bed.gz</span></code></pre></div>
<p>Then using the <code>iscream.paper</code> package at <a href="https://github.com/huishenlab/iscream.paper" class="external-link uri">https://github.com/huishenlab/iscream.paper</a> we ran the
benchmarks and produced the figures.</p>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-krueger2011" class="csl-entry">
Krueger, Felix, and Simon R. Andrews. 2011. <span>“Bismark: A Flexible
Aligner and Methylation Caller for <span>Bisulfite-Seq</span>
Applications.”</span> <em>Bioinformatics</em> 27 (11): 1571–72. <a href="https://doi.org/10.1093/bioinformatics/btr167" class="external-link">https://doi.org/10.1093/bioinformatics/btr167</a>.
</div>
<div id="ref-luo2018a" class="csl-entry">
Luo, Chongyuan, Angeline Rivkin, Jingtian Zhou, Justin P. Sandoval,
Laurie Kurihara, Jacinta Lucero, Rosa Castanon, et al. 2018.
<span>“Robust Single-Cell <span>DNA</span> Methylome Profiling with
<span class="nocase">snmC-seq2</span>.”</span> <em>Nat Commun</em> 9
(1): 3824. <a href="https://doi.org/10.1038/s41467-018-06355-2" class="external-link">https://doi.org/10.1038/s41467-018-06355-2</a>.
</div>
<div id="ref-zhou2024" class="csl-entry">
Zhou, Wanding, Benjamin K Johnson, Jacob Morrison, Ian Beddows, James
Eapen, Efrat Katsman, Ayush Semwal, et al. 2024.
<span>“<span>BISCUIT</span>: An Efficient, Standards-Compliant Tool
Suite for Simultaneous Genetic and Epigenetic Inference in Bulk and
Single-Cell Studies.”</span> <em>Nucleic Acids Research</em> 52 (6):
gkae097. <a href="https://doi.org/10.1093/nar/gkae097" class="external-link">https://doi.org/10.1093/nar/gkae097</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by James Eapen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
