<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Query records from tabixed BED files — tabix • iscream</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Query records from tabixed BED files — tabix"><meta name="description" content="Query records from tabixed BED files"><meta property="og:description" content="Query records from tabixed BED files"><meta property="og:image" content="https://huishenlab.github.io/iscream/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">iscream</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.9</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/iscream.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/performance.html">Improving iscream performance</a></li>
    <li><a class="dropdown-item" href="../articles/data_structures.html">iscream compatible data structures</a></li>
    <li><a class="dropdown-item" href="../articles/TSS.html">Plotting TSS methylation profiles</a></li>
    <li><a class="dropdown-item" href="../articles/tabix.html">iscream vs Rsamtools::scanTabix</a></li>
    <li><a class="dropdown-item" href="../articles/htslib.html">Getting htslib headers</a></li>
    <li><a class="dropdown-item" href="../articles/manuscript_data.html">Manuscript data availabiliy</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/huishenlab/iscream/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Query records from tabixed BED files</h1>
      <small class="dont-index">Source: <a href="https://github.com/huishenlab/iscream/blob/main/R/tabix.R" class="external-link"><code>R/tabix.R</code></a>, <a href="https://github.com/huishenlab/iscream/blob/main/R/tabix_gr.R" class="external-link"><code>R/tabix_gr.R</code></a>, <a href="https://github.com/huishenlab/iscream/blob/main/R/tabix_raw.R" class="external-link"><code>R/tabix_raw.R</code></a></small>
      <div class="d-none name"><code>tabix.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Query records from tabixed BED files</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">tabix</span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span>, aligner <span class="op">=</span> <span class="cn">NULL</span>, col.names <span class="op">=</span> <span class="cn">NULL</span>, nthreads <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">tabix_gr</span><span class="op">(</span></span>
<span>  <span class="va">bedfiles</span>,</span>
<span>  <span class="va">regions</span>,</span>
<span>  aligner <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  col.names <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  zero_based <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  nthreads <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">tabix_raw</span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span>, nthreads <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-bedfiles">bedfiles<a class="anchor" aria-label="anchor" href="#arg-bedfiles"></a></dt>
<dd><p>The BED files to be queried</p></dd>


<dt id="arg-regions">regions<a class="anchor" aria-label="anchor" href="#arg-regions"></a></dt>
<dd><p>A vector, data frame or GenomicRanges of genomic regions. See
details.</p></dd>


<dt id="arg-aligner">aligner<a class="anchor" aria-label="anchor" href="#arg-aligner"></a></dt>
<dd><p>The aligner used to produce the BED files - one of "biscuit",
"bismark", "bsbolt". Will set the result data.table's column names based on
this argument.</p></dd>


<dt id="arg-col-names">col.names<a class="anchor" aria-label="anchor" href="#arg-col-names"></a></dt>
<dd><p>A vector of column names for the data columns of the
result.table, not including "chr", "start", and "end". Set if your BED file
is not from the supported aligners or is a general BED file.</p></dd>


<dt id="arg-nthreads">nthreads<a class="anchor" aria-label="anchor" href="#arg-nthreads"></a></dt>
<dd><p>Set number of threads to use overriding the
<code>"iscream.threads"</code> option. See <code><a href="set_threads.html">?set_threads</a></code> for more information.</p></dd>


<dt id="arg-zero-based">zero_based<a class="anchor" aria-label="anchor" href="#arg-zero-based"></a></dt>
<dd><p>Whether the input BED file has a zero-based start column -
used when coverting the result data frame to GenomicRanges.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>

<ul><li><p><code>tabix()</code>: A data frame</p></li>
<li><p><code>tabix_gr()</code>: A <code>GRanges</code> object for single files and <code>GRangesList</code> for
multiple files. When making <code>GRanges</code>, the 0-based records from BED-files
will be converted to 1-based with
<code><a href="https://rdrr.io/pkg/GenomicRanges/man/makeGRangesFromDataFrame.html" class="external-link">GenomicRanges::makeGRangesFromDataFrame()</a></code>. Bismark's coverage files will
not be converted as they are already 1-based and the <code>ranges</code> slot will be
only one position.</p></li>
<li><p><code>tabix_raw()</code>: A named list of raw strings from the regions in the style
of <code><a href="https://rdrr.io/pkg/Rsamtools/man/scanTabix.html" class="external-link">Rsamtools::scanTabix</a></code></p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>

<div class="section">
<h3 id="query-method">Query method<a class="anchor" aria-label="anchor" href="#query-method"></a></h3>


<p>'<em>iscream</em> has two methods to query records from BED files:</p><ul><li><p>the <em>tabix</em> shell executable: fast since its output can be redirected to a
file (which <code><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">data.table::fread()</a></code> can then read) instead of having to
allocate memory and store it during the query</p></li>
<li><p><em>iscream's</em> tabix implementation, based on the <em>tabix</em> executable using
<em>htslib</em>, but slower on large queries since it stores the records as they
are found instead of writing to a file. However it's able to store each
region's records independently instead of in a single file and is used in
<code><a href="make_mat.html">make_mat()</a></code>, <code><a href="make_mat_bsseq.html">make_mat_bsseq()</a></code>, and <code><a href="summarize_regions.html">summarize_regions()</a></code>.</p></li>
</ul><p>When <em>iscream</em> is attached, it checks that the <em>tabix</em> executable is
available with <code><a href="https://rdrr.io/r/base/Sys.which.html" class="external-link">Sys.which()</a></code> and, if available, sets <code>options("tabix.method" = "shell")</code>. <code>tabix()</code> then uses the <em>tabix</em> executable to make queries,
except for <code>tabix_raw()</code>. If <em>tabix</em> is not found, <em>iscream</em> uses its tabix
implementation. To use only <em>iscream's</em> tabix implementation, set
<code>options("tabix.method" = "htslib")</code>.</p>
</div>

<div class="section">
<h3 id="input-region-formats">Input region formats<a class="anchor" aria-label="anchor" href="#input-region-formats"></a></h3>


<p>The input regions format may be string vector in the form "chr:start-end",
a dataframe with "chr", "start" and "end" columns or a <code>GRanges</code> object.
Input regions must be 1-based. When using <code>"htslib"</code> as the query method, if
the input <code>GRanges</code> object of regions contains any single locus regions
where the start and end positions are the same, iscream will notify that
such regions were found and fixed as <code>chr:start</code> format strings are invalid
for the htslib API (see <code><a href="get_granges_string.html">?get_granges_string</a></code>).</p>
</div>

    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">bedfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"iscream"</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"[a|b|c|d].bed.gz$"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1:1-6"</span>, <span class="st">"chr1:7-10"</span>, <span class="st">"chr1:11-14"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">tabix</span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span>, col.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"beta"</span>, <span class="st">"coverage"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        chr start   end  beta coverage   file</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     &lt;char&gt; &lt;int&gt; &lt;int&gt; &lt;num&gt;    &lt;int&gt; &lt;char&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  1:   chr1     0     2   1.0        1      a</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  2:   chr1     2     4   1.0        1      a</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  3:   chr1     4     6   0.0        2      a</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  4:   chr1     6     8   0.0        1      a</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  5:   chr1     8    10   0.5        2      a</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  6:   chr1    10    12   1.0        2      a</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  7:   chr1    12    14   1.0        3      a</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  8:   chr1     0     2   0.0        2      b</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  9:   chr1     4     6   1.0        2      b</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10:   chr1     6     8   1.0        1      b</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11:   chr1    10    12   0.0        2      b</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12:   chr1    12    14   1.0        1      b</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13:   chr1     2     4   1.0        2      c</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14:   chr1     6     8   0.0        2      c</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15:   chr1     8    10   1.0        1      c</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16:   chr1     0     2   1.0        1      d</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17:   chr1     2     4   1.0        2      d</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18:   chr1     6     8   0.0        1      d</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 19:   chr1     8    10   0.5        2      d</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 20:   chr1    12    14   1.0        1      d</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        chr start   end  beta coverage   file</span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">"GenomicRanges"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="fu">tabix_gr</span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span>, col.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"beta"</span>, <span class="st">"coverage"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘generics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     as.difftime, as.factor, as.ordered, intersect, is.element, setdiff,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     setequal, union</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘BiocGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:stats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     IQR, mad, sd, var, xtabs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Filter, Find, Map, Position, Reduce, anyDuplicated, aperm, append,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     as.data.frame, basename, cbind, colnames, dirname, do.call,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     duplicated, eval, evalq, get, grep, grepl, is.unsorted, lapply,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     mapply, match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rank, rbind, rownames, sapply, saveRDS, table, tapply, unique,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     unsplit, which.max, which.min</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘S4Vectors’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:data.table’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     first, second</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:utils’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     findMatches</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     I, expand.grid, unname</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘IRanges’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:data.table’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     shift</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRangesList object of length 4:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $a</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 7 ranges and 3 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames    ranges strand |      beta  coverage        file</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;integer&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]     chr1       1-2      * |       1.0         1           a</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]     chr1       3-4      * |       1.0         1           a</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3]     chr1       5-6      * |       0.0         2           a</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [4]     chr1       7-8      * |       0.0         1           a</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [5]     chr1      9-10      * |       0.5         2           a</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [6]     chr1     11-12      * |       1.0         2           a</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [7]     chr1     13-14      * |       1.0         3           a</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $b</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 5 ranges and 3 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames    ranges strand |      beta  coverage        file</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;integer&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]     chr1       1-2      * |         0         2           b</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]     chr1       5-6      * |         1         2           b</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3]     chr1       7-8      * |         1         1           b</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [4]     chr1     11-12      * |         0         2           b</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [5]     chr1     13-14      * |         1         1           b</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $c</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 3 ranges and 3 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames    ranges strand |      beta  coverage        file</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;integer&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]     chr1       3-4      * |         1         2           c</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]     chr1       7-8      * |         0         2           c</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3]     chr1      9-10      * |         1         1           c</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $d</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 5 ranges and 3 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames    ranges strand |      beta  coverage        file</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;integer&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]     chr1       1-2      * |       1.0         1           d</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]     chr1       3-4      * |       1.0         2           d</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3]     chr1       7-8      * |       0.0         1           d</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [4]     chr1      9-10      * |       0.5         2           d</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [5]     chr1     13-14      * |       1.0         1           d</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu">tabix_raw</span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $a</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $a$`chr1:1-6`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "chr1\t0\t2\t1.000\t1" "chr1\t2\t4\t1.000\t1" "chr1\t4\t6\t0.000\t2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $a$`chr1:7-10`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "chr1\t6\t8\t0.000\t1"  "chr1\t8\t10\t0.500\t2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $a$`chr1:11-14`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "chr1\t10\t12\t1.000\t2" "chr1\t12\t14\t1.000\t3"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $b</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $b$`chr1:1-6`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "chr1\t0\t2\t0.000\t2" "chr1\t4\t6\t1.000\t2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $b$`chr1:7-10`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "chr1\t6\t8\t1.000\t1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $b$`chr1:11-14`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "chr1\t10\t12\t0.000\t2" "chr1\t12\t14\t1.000\t1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $c</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $c$`chr1:1-6`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "chr1\t2\t4\t1.000\t2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $c$`chr1:7-10`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "chr1\t6\t8\t0.000\t2"  "chr1\t8\t10\t1.000\t1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $c$`chr1:11-14`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> character(0)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $d</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $d$`chr1:1-6`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "chr1\t0\t2\t1.000\t1" "chr1\t2\t4\t1.000\t2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $d$`chr1:7-10`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "chr1\t6\t8\t0.000\t1"  "chr1\t8\t10\t0.500\t2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $d$`chr1:11-14`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "chr1\t12\t14\t1.000\t1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by James Eapen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

