<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Apply a function over BED file records within genomic features — Cpp_summarize_regions • iscream</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Apply a function over BED file records within genomic features — Cpp_summarize_regions"><meta name="description" content="This function should be called from summarize_regions() since there are few
sanity checks on the C++ side."><meta property="og:description" content="This function should be called from summarize_regions() since there are few
sanity checks on the C++ side."><meta property="og:image" content="https://huishenlab.github.io/iscream/logo.png"><meta name="robots" content="noindex"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">iscream</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/iscream.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/performance.html">Improving iscream performance</a></li>
    <li><a class="dropdown-item" href="../articles/data_structures.html">iscream compatible data structures</a></li>
    <li><a class="dropdown-item" href="../articles/TSS.html">Plotting TSS methylation profiles</a></li>
    <li><a class="dropdown-item" href="../articles/tabix.html">iscream vs Rsamtools::scanTabix</a></li>
    <li><a class="dropdown-item" href="../articles/htslib.html">Getting htslib headers</a></li>
    <li><a class="dropdown-item" href="../articles/manuscript_data.html">Manuscript data availabiliy</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/huishenlab/iscream/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Apply a function over BED file records within genomic features</h1>
      <small class="dont-index">Source: <a href="https://github.com/huishenlab/iscream/blob/dev/R/RcppExports.R" class="external-link"><code>R/RcppExports.R</code></a></small>
      <div class="d-none name"><code>Cpp_summarize_regions.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function should be called from <code><a href="summarize_regions.html">summarize_regions()</a></code> since there are few
sanity checks on the C++ side.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">Cpp_summarize_regions</span><span class="op">(</span></span>
<span>  <span class="va">bedfiles</span>,</span>
<span>  <span class="va">regions</span>,</span>
<span>  <span class="va">fun_vec</span>,</span>
<span>  <span class="va">col_indices</span>,</span>
<span>  <span class="va">col_names</span>,</span>
<span>  <span class="va">aligner</span>,</span>
<span>  mval <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  region_rownames <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  nthreads <span class="op">=</span> <span class="fl">1L</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-bedfiles">bedfiles<a class="anchor" aria-label="anchor" href="#arg-bedfiles"></a></dt>
<dd><p>A vector of BED file paths</p></dd>


<dt id="arg-regions">regions<a class="anchor" aria-label="anchor" href="#arg-regions"></a></dt>
<dd><p>A vector of genomic regions</p></dd>


<dt id="arg-fun-vec">fun_vec<a class="anchor" aria-label="anchor" href="#arg-fun-vec"></a></dt>
<dd><p>Vector of the armadillo-supported stats functions to apply over the
CpGs in the ' regions: <code>"sum"</code>, <code>"mean"</code>, <code>"median"</code>, <code>"stddev"</code>,
<code>"variance"</code> "<code>count</code>", <code>"min"</code>,<code>"max"</code>, and <code>"range"</code>.</p></dd>


<dt id="arg-col-indices">col_indices<a class="anchor" aria-label="anchor" href="#arg-col-indices"></a></dt>
<dd><p>A vector of genomic regions</p></dd>


<dt id="arg-col-names">col_names<a class="anchor" aria-label="anchor" href="#arg-col-names"></a></dt>
<dd><p>A vector of genomic regions</p></dd>


<dt id="arg-mval">mval<a class="anchor" aria-label="anchor" href="#arg-mval"></a></dt>
<dd><p>Calculates M values when TRUE, use beta values when FALSE</p></dd>


<dt id="arg-region-rownames">region_rownames<a class="anchor" aria-label="anchor" href="#arg-region-rownames"></a></dt>
<dd><p>Whether to set rownames to the regions strings. Not
necessary if your regions vector is unnamed. If its names, then the "feature"
column is set to the names and the rownames are set to the regions string</p></dd>


<dt id="arg-nthreads">nthreads<a class="anchor" aria-label="anchor" href="#arg-nthreads"></a></dt>
<dd><p>Number of cores to use. See details.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A summary data.frame</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The optimal number of threads depends on the number of bedfiles, but is set
to half the available OpenMP cores. See <code><a href="get_threads.html">?get_threads</a></code> for more details. It
can be manaully set with <code><a href="set_threads.html">set_threads()</a></code>.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by James Eapen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

