<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Summarize CpGs methylation information over genomic regions — summarize_meth_regions • iscream</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Summarize CpGs methylation information over genomic regions — summarize_meth_regions"><meta name="description" content='Run summarizing functions on the CpGs in BED files across genomic regions.
Parallelized across files using threads from the "iscream.threads" option.'><meta property="og:description" content='Run summarizing functions on the CpGs in BED files across genomic regions.
Parallelized across files using threads from the "iscream.threads" option.'><meta property="og:image" content="https://huishenlab.github.io/iscream/logo.png"><meta name="robots" content="noindex"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">iscream</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/iscream.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/performance.html">Improving iscream performance</a></li>
    <li><a class="dropdown-item" href="../articles/data_structures.html">iscream compatible data structures</a></li>
    <li><a class="dropdown-item" href="../articles/TSS.html">Plotting TSS methylation profiles</a></li>
    <li><a class="dropdown-item" href="../articles/tabix.html">iscream vs Rsamtools::scanTabix</a></li>
    <li><a class="dropdown-item" href="../articles/htslib.html">Getting htslib headers</a></li>
    <li><a class="dropdown-item" href="../articles/manuscript_data.html">Manuscript data availabiliy</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/huishenlab/iscream/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Summarize CpGs methylation information over genomic regions</h1>
      <small class="dont-index">Source: <a href="https://github.com/huishenlab/iscream/blob/dev/R/summarize_meth_regions.R" class="external-link"><code>R/summarize_meth_regions.R</code></a></small>
      <div class="d-none name"><code>summarize_meth_regions.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Run summarizing functions on the CpGs in BED files across genomic regions.
Parallelized across files using threads from the <code>"iscream.threads"</code> option.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">summarize_meth_regions</span><span class="op">(</span></span>
<span>  <span class="va">bedfiles</span>,</span>
<span>  <span class="va">regions</span>,</span>
<span>  fun <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  aligner <span class="op">=</span> <span class="st">"biscuit"</span>,</span>
<span>  feature_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  mval <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  set_region_rownames <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  nthreads <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-bedfiles">bedfiles<a class="anchor" aria-label="anchor" href="#arg-bedfiles"></a></dt>
<dd><p>A vector of BED file paths</p></dd>


<dt id="arg-regions">regions<a class="anchor" aria-label="anchor" href="#arg-regions"></a></dt>
<dd><p>A vector, data frame or GenomicRanges of genomic regions. See
details.</p></dd>


<dt id="arg-fun">fun<a class="anchor" aria-label="anchor" href="#arg-fun"></a></dt>
<dd><p>Function(s) to apply over the region. See details.</p></dd>


<dt id="arg-aligner">aligner<a class="anchor" aria-label="anchor" href="#arg-aligner"></a></dt>
<dd><p>The aligner used to produce the BED files - one of "biscuit",
"bismark", "bsbolt".</p></dd>


<dt id="arg-feature-col">feature_col<a class="anchor" aria-label="anchor" href="#arg-feature-col"></a></dt>
<dd><p>Column name of the input <code>regions</code> data frame containing
a name for each genomic region. Set only if the using a data frame as the
input regions format. See details.</p></dd>


<dt id="arg-mval">mval<a class="anchor" aria-label="anchor" href="#arg-mval"></a></dt>
<dd><p>Whether to calculate the M value (coverage \(\times \beta\))
or use the beta value when applying the function.</p></dd>


<dt id="arg-set-region-rownames">set_region_rownames<a class="anchor" aria-label="anchor" href="#arg-set-region-rownames"></a></dt>
<dd><p>Use the region strings as the returned data
frame's rownames. Can be useful if you have a named regions and want both
the regions strings rownames and the feature names. See details.</p></dd>


<dt id="arg-nthreads">nthreads<a class="anchor" aria-label="anchor" href="#arg-nthreads"></a></dt>
<dd><p>Set number of threads to use overriding the
<code>"iscream.threads"</code> option. See <code><a href="set_threads.html">?set_threads</a></code> for more information.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data.frame</p>
    </div>
    <div class="section level2">
    <h2 id="supported-functions">Supported functions<a class="anchor" aria-label="anchor" href="#supported-functions"></a></h2>

<ul><li><p>Sum: <code>"sum"</code></p></li>
<li><p>Mean: <code>"mean"</code></p></li>
<li><p>Median: <code>"median"</code></p></li>
<li><p>Standard deviation: <code>"stddev"</code></p></li>
<li><p>Variance: <code>"variance"</code></p></li>
<li><p>Minimum: <code>"min"</code></p></li>
<li><p>Maximum: <code>"max"</code></p></li>
<li><p>Range: <code>"range"</code></p></li>
<li><p>No. of records in the region: <code>"count"</code></p></li>
</ul><p>The summarizing computations are backed by the Armadillo library. See
<a href="https://arma.sourceforge.net/docs.html#stats_fns" class="external-link">https://arma.sourceforge.net/docs.html#stats_fns</a> for futher details on the
supported functions</p>
    </div>
    <div class="section level2">
    <h2 id="using-feature-identifiers">Using feature identifiers<a class="anchor" aria-label="anchor" href="#using-feature-identifiers"></a></h2>
    <p><code>regions</code> may be string vector in the form "chr:start-end", a GRanges
object or a data frame with "chr", "start", and "end" columns. The <code>feature</code>
column of the output will contain a "chr:start-end" identifier for each
summarized region. To use other identifiers, like a gene name for a region
instead of the coordinates, set the names of the vector or GRanges to those
identifiers. These names will be used instead of the genomic region string
to describe each feature in the output dataframe. If <code>regions</code> is a data
frame make an additional column with the identifiers and pass that column
name to <code>feature_col</code>. See examples.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># also see examples from ?summarize_regions</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">bedfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"iscream"</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"[a|b|c|d].bed.gz$"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># make a vector of regions</span></span></span>
<span class="r-in"><span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1:1-6"</span>, <span class="st">"chr1:7-10"</span>, <span class="st">"chr1:11-14"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">summarize_meth_regions</span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13:28:38.472042] [iscream::summarize_regions] [info] Summarizing 3 regions from 4 bedfiles</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13:28:38.472065] [iscream::summarize_regions] [info] using sum, mean, median, stddev, variance, min, max, range, count</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13:28:38.472069] [iscream::summarize_regions] [info] with columns 4, 5 as coverage, M</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       feature file coverage.sum M.sum coverage.mean    M.mean coverage.median</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    chr1:1-6    a            4     2      1.333333 0.6666667             1.0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   chr1:7-10    a            3     1      1.500000 0.5000000             1.5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  chr1:11-14    a            5     5      2.500000 2.5000000             2.5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4    chr1:1-6    b            4     2      2.000000 1.0000000             2.0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   chr1:7-10    b            1     1      1.000000 1.0000000             1.0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6  chr1:11-14    b            3     1      1.500000 0.5000000             1.5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7    chr1:1-6    c            2     2      2.000000 2.0000000             2.0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8   chr1:7-10    c            3     1      1.500000 0.5000000             1.5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9  chr1:11-14    c           NA    NA            NA        NA              NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10   chr1:1-6    d            3     3      1.500000 1.5000000             1.5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11  chr1:7-10    d            3     1      1.500000 0.5000000             1.5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12 chr1:11-14    d            1     1      1.000000 1.0000000             1.0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    M.median coverage.stddev  M.stddev coverage.variance M.variance coverage.min</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       1.0       0.5773503 0.5773503         0.3333333  0.3333333            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       0.5       0.7071068 0.7071068         0.5000000  0.5000000            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       2.5       0.7071068 0.7071068         0.5000000  0.5000000            2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       1.0       0.0000000 1.4142136         0.0000000  2.0000000            2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5       1.0       0.0000000 0.0000000         0.0000000  0.0000000            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6       0.5       0.7071068 0.7071068         0.5000000  0.5000000            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7       2.0       0.0000000 0.0000000         0.0000000  0.0000000            2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8       0.5       0.7071068 0.7071068         0.5000000  0.5000000            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9        NA              NA        NA                NA         NA           NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10      1.5       0.7071068 0.7071068         0.5000000  0.5000000            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11      0.5       0.7071068 0.7071068         0.5000000  0.5000000            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12      1.0       0.0000000 0.0000000         0.0000000  0.0000000            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    M.min coverage.max M.max coverage.range M.range cpg_count</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1      0            2     1              1       1         3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2      0            2     1              1       1         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3      2            3     3              1       1         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4      0            2     2              0       2         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5      1            1     1              0       0         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6      0            2     1              1       1         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7      2            2     2              0       0         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8      0            2     1              1       1         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9     NA           NA    NA             NA      NA        NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10     1            2     2              1       1         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11     0            2     1              1       1         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12     1            1     1              0       0         1</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">regions</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">summarize_meth_regions</span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span>, fun <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mean"</span>, <span class="st">"stddev"</span><span class="op">)</span>, mval <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13:28:38.490191] [iscream::summarize_regions] [info] Summarizing 3 regions from 4 bedfiles</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13:28:38.490204] [iscream::summarize_regions] [info] using mean, stddev</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13:28:38.490207] [iscream::summarize_regions] [info] with columns 4, 5 as coverage, beta</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    feature file coverage.mean beta.mean coverage.stddev beta.stddev</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1        A    a      1.333333 0.6666667       0.5773503   0.5773503</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2        B    a      1.500000 0.2500000       0.7071068   0.3535534</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3        C    a      2.500000 1.0000000       0.7071068   0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4        A    b      2.000000 0.5000000       0.0000000   0.7071068</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5        B    b      1.000000 1.0000000       0.0000000   0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6        C    b      1.500000 0.5000000       0.7071068   0.7071068</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7        A    c      2.000000 1.0000000       0.0000000   0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8        B    c      1.500000 0.5000000       0.7071068   0.7071068</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9        C    c            NA        NA              NA          NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10       A    d      1.500000 1.0000000       0.7071068   0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11       B    d      1.500000 0.2500000       0.7071068   0.3535534</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12       C    d      1.000000 1.0000000       0.0000000   0.0000000</span>
<span class="r-in"><span><span class="fu">summarize_meth_regions</span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span>, fun <span class="op">=</span> <span class="st">"sum"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13:28:38.494359] [iscream::summarize_regions] [info] Summarizing 3 regions from 4 bedfiles</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13:28:38.494370] [iscream::summarize_regions] [info] using sum</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13:28:38.494373] [iscream::summarize_regions] [info] with columns 4, 5 as coverage, M</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    feature file coverage.sum M.sum</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1        A    a            4     2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2        B    a            3     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3        C    a            5     5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4        A    b            4     2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5        B    b            1     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6        C    b            3     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7        A    c            2     2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8        B    c            3     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9        C    c           NA    NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10       A    d            3     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11       B    d            3     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12       C    d            1     1</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by James Eapen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

