<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>iscream compatible data structures • iscream</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="iscream compatible data structures">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">iscream</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/iscream.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/performance.html">Improving iscream performance</a></li>
    <li><a class="dropdown-item" href="../articles/data_structures.html">iscream compatible data structures</a></li>
    <li><a class="dropdown-item" href="../articles/TSS.html">Plotting TSS methylation profiles</a></li>
    <li><a class="dropdown-item" href="../articles/tabix.html">iscream vs Rsamtools::scanTabix</a></li>
    <li><a class="dropdown-item" href="../articles/htslib.html">Getting htslib headers</a></li>
    <li><a class="dropdown-item" href="../articles/manuscript_data.html">Manuscript data availabiliy</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/huishenlab/iscream/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>iscream compatible data structures</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/huishenlab/iscream/blob/main/vignettes/data_structures.Rmd" class="external-link"><code>vignettes/data_structures.Rmd</code></a></small>
      <div class="d-none name"><code>data_structures.Rmd</code></div>
    </div>

    
    
<p>The examples here show how iscream output can be converted into other
data structures for further analysis.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://huishenlab.github.io/iscream/">iscream</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## iscream using 1 thread by default but parallelly::availableCores() detects 4 possibly available threads. See `?set_threads` for information on multithreading before trying to use more.</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"iscream"</span><span class="op">)</span></span>
<span><span class="va">bedfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">data_dir</span>, pattern <span class="op">=</span> <span class="st">"[a|b|c|d].bed.gz$"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>A <span class="op">=</span> <span class="st">"chr1:1-6"</span>, B <span class="op">=</span> <span class="st">"chr1:7-10"</span>, C <span class="op">=</span> <span class="st">"chr1:11-14"</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="genomicranges">
<em><a href="https://bioconductor.org/packages/3.21/GenomicRanges" class="external-link">GenomicRanges</a></em><a class="anchor" aria-label="anchor" href="#genomicranges"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">"GenomicRanges"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"The 'GenomicRanges' package must be installed for this functionality"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><code>GRanges</code> objects can be used as the input regions to all
of iscream’s functions and can be returned by <code><a href="../reference/tabix.html">tabix_gr()</a></code>
and <code><a href="../reference/make_mat.html">make_mat_gr()</a></code>.</p>
<div class="section level3">
<h3 id="from-tabix-queries">From <code>tabix</code> queries<a class="anchor" aria-label="anchor" href="#from-tabix-queries"></a>
</h3>
<p><code><a href="../reference/tabix.html">tabix_gr()</a></code> returns a <code>GenomicRanges</code> object.
The <code>regions</code> parameter can be a string vector, a data frame
or a <code>GRanges</code> object. If given a <code>GRanges</code> object
with metadata, those columns will be preserved in the output.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="va">regions</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">values</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></span><span class="op">(</span></span>
<span>    gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span>, <span class="st">"gene3"</span><span class="op">)</span>,</span>
<span>    some_metadata <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"s1"</span>, <span class="st">"s2"</span>, <span class="st">"s3"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">gr</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 2 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand |        gene some_metadata</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;   &lt;character&gt;</span></span>
<span><span class="co">##   A     chr1       1-6      * |       gene1            s1</span></span>
<span><span class="co">##   B     chr1      7-10      * |       gene2            s2</span></span>
<span><span class="co">##   C     chr1     11-14      * |       gene3            s3</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tabix.html">tabix_gr</a></span><span class="op">(</span><span class="va">bedfiles</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">gr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 7 ranges and 4 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |        V1        V2        gene some_metadata</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;integer&gt; &lt;character&gt;   &lt;character&gt;</span></span>
<span><span class="co">##   [1]     chr1       1-2      * |       1.0         1       gene1            s1</span></span>
<span><span class="co">##   [2]     chr1       3-4      * |       1.0         1       gene1            s1</span></span>
<span><span class="co">##   [3]     chr1       5-6      * |       0.0         2       gene1            s1</span></span>
<span><span class="co">##   [4]     chr1       7-8      * |       0.0         1       gene2            s2</span></span>
<span><span class="co">##   [5]     chr1      9-10      * |       0.5         2       gene2            s2</span></span>
<span><span class="co">##   [6]     chr1     11-12      * |       1.0         2       gene3            s3</span></span>
<span><span class="co">##   [7]     chr1     13-14      * |       1.0         3       gene3            s3</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>The <code>data.table</code> output of <code><a href="../reference/tabix.html">tabix()</a></code> can also
be piped into <code>GRanges</code>, but does not preserve input
metadata.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tabix.html">tabix</a></span><span class="op">(</span><span class="va">bedfiles</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">gr</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/makeGRangesFromDataFrame.html" class="external-link">makeGRangesFromDataFrame</a></span><span class="op">(</span></span>
<span>    starts.in.df.are.0based <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    keep.extra.columns <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 7 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |        V1        V2</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">##   [1]     chr1       1-2      * |       1.0         1</span></span>
<span><span class="co">##   [2]     chr1       3-4      * |       1.0         1</span></span>
<span><span class="co">##   [3]     chr1       5-6      * |       0.0         2</span></span>
<span><span class="co">##   [4]     chr1       7-8      * |       0.0         1</span></span>
<span><span class="co">##   [5]     chr1      9-10      * |       0.5         2</span></span>
<span><span class="co">##   [6]     chr1     11-12      * |       1.0         2</span></span>
<span><span class="co">##   [7]     chr1     13-14      * |       1.0         3</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>If the input BED file is not zero-based (e.g. Bismark coverage
files), set <code>zero_based = FALSE</code> in the <code><a href="../reference/tabix.html">tabix()</a></code>
call to get the correct conversion from data frame to GenomicRanges.</p>
</div>
<div class="section level3">
<h3 id="from-summarize_regions">From <code>summarize_regions()</code><a class="anchor" aria-label="anchor" href="#from-summarize_regions"></a>
</h3>
<p><code><a href="../reference/summarize_regions.html">summarize_regions()</a></code> returns a data frame with a feature
column identifying each summary row’s genomic region.</p>
<p>If the region features are not named (see
<code><a href="../reference/summarize_regions.html">?summarize_regions</a></code>), pass the <code>feature</code> column
with the genomic regions as the GRanges input regions. Here, since the
<code>regions</code> vector is named, using <code>unname</code> will
cause the <code>feature</code> column to populate with regions
strings:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">summary</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_meth_regions.html">summarize_meth_regions</a></span><span class="op">(</span></span>
<span>  <span class="va">bedfiles</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unname.html" class="external-link">unname</a></span><span class="op">(</span><span class="va">regions</span><span class="op">)</span>,</span>
<span>  fun <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sum"</span>, <span class="st">"mean"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [16:33:44.236529] [iscream::summarize_regions] [info] Summarizing 3 regions from 4 bedfiles</span></span>
<span><span class="co">## [16:33:44.236568] [iscream::summarize_regions] [info] using sum, mean</span></span>
<span><span class="co">## [16:33:44.236573] [iscream::summarize_regions] [info] with columns 4, 5 as coverage, M</span></span></code></pre>
<pre><code><span><span class="co">##       feature file coverage.sum M.sum coverage.mean    M.mean</span></span>
<span><span class="co">## 1    chr1:1-6    a            4     2      1.333333 0.6666667</span></span>
<span><span class="co">## 2   chr1:7-10    a            3     1      1.500000 0.5000000</span></span>
<span><span class="co">## 3  chr1:11-14    a            5     5      2.500000 2.5000000</span></span>
<span><span class="co">## 4    chr1:1-6    b            4     2      2.000000 1.0000000</span></span>
<span><span class="co">## 5   chr1:7-10    b            1     1      1.000000 1.0000000</span></span>
<span><span class="co">## 6  chr1:11-14    b            3     1      1.500000 0.5000000</span></span>
<span><span class="co">## 7    chr1:1-6    c            2     2      2.000000 2.0000000</span></span>
<span><span class="co">## 8   chr1:7-10    c            3     1      1.500000 0.5000000</span></span>
<span><span class="co">## 9  chr1:11-14    c           NA    NA            NA        NA</span></span>
<span><span class="co">## 10   chr1:1-6    d            3     3      1.500000 1.5000000</span></span>
<span><span class="co">## 11  chr1:7-10    d            3     1      1.500000 0.5000000</span></span>
<span><span class="co">## 12 chr1:11-14    d            1     1      1.000000 1.0000000</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="va">summary</span><span class="op">$</span><span class="va">feature</span>, summary <span class="op">=</span> <span class="va">summary</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 12 ranges and 5 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand | summary.file summary.coverage.sum</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;character&gt;            &lt;numeric&gt;</span></span>
<span><span class="co">##    [1]     chr1       1-6      * |            a                    4</span></span>
<span><span class="co">##    [2]     chr1      7-10      * |            a                    3</span></span>
<span><span class="co">##    [3]     chr1     11-14      * |            a                    5</span></span>
<span><span class="co">##    [4]     chr1       1-6      * |            b                    4</span></span>
<span><span class="co">##    [5]     chr1      7-10      * |            b                    1</span></span>
<span><span class="co">##    ...      ...       ...    ... .          ...                  ...</span></span>
<span><span class="co">##    [8]     chr1      7-10      * |            c                    3</span></span>
<span><span class="co">##    [9]     chr1     11-14      * |            c                   NA</span></span>
<span><span class="co">##   [10]     chr1       1-6      * |            d                    3</span></span>
<span><span class="co">##   [11]     chr1      7-10      * |            d                    3</span></span>
<span><span class="co">##   [12]     chr1     11-14      * |            d                    1</span></span>
<span><span class="co">##        summary.M.sum summary.coverage.mean summary.M.mean</span></span>
<span><span class="co">##            &lt;numeric&gt;             &lt;numeric&gt;      &lt;numeric&gt;</span></span>
<span><span class="co">##    [1]             2               1.33333       0.666667</span></span>
<span><span class="co">##    [2]             1               1.50000       0.500000</span></span>
<span><span class="co">##    [3]             5               2.50000       2.500000</span></span>
<span><span class="co">##    [4]             2               2.00000       1.000000</span></span>
<span><span class="co">##    [5]             1               1.00000       1.000000</span></span>
<span><span class="co">##    ...           ...                   ...            ...</span></span>
<span><span class="co">##    [8]             1                   1.5            0.5</span></span>
<span><span class="co">##    [9]            NA                    NA             NA</span></span>
<span><span class="co">##   [10]             3                   1.5            1.5</span></span>
<span><span class="co">##   [11]             1                   1.5            0.5</span></span>
<span><span class="co">##   [12]             1                   1.0            1.0</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>If the input regions are named use
<code>set_region_rownames = TRUE</code> so that the genomic regions
strings are preserved and use them as the <code>GRanges</code> input
regions.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">summary</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_regions.html">summarize_regions</a></span><span class="op">(</span></span>
<span>  <span class="va">bedfiles</span>,</span>
<span>  <span class="va">regions</span>,</span>
<span>  column <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  set_region_rownames <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  fun <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sum"</span>, <span class="st">"mean"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [16:33:44.359038] [iscream::summarize_regions] [info] Summarizing 3 regions from 4 bedfiles</span></span>
<span><span class="co">## [16:33:44.359067] [iscream::summarize_regions] [info] using sum, mean</span></span>
<span><span class="co">## [16:33:44.359072] [iscream::summarize_regions] [info] with columns 4 as V1</span></span></code></pre>
<pre><code><span><span class="co">##            feature file V1.sum   V1.mean</span></span>
<span><span class="co">## chr1:1-6         A    a    2.0 0.6666667</span></span>
<span><span class="co">## chr1:7-10        B    a    0.5 0.2500000</span></span>
<span><span class="co">## chr1:11-14       C    a    2.0 1.0000000</span></span>
<span><span class="co">## chr1:1-6         A    b    1.0 0.5000000</span></span>
<span><span class="co">## chr1:7-10        B    b    1.0 1.0000000</span></span>
<span><span class="co">## chr1:11-14       C    b    1.0 0.5000000</span></span>
<span><span class="co">## chr1:1-6         A    c    1.0 1.0000000</span></span>
<span><span class="co">## chr1:7-10        B    c    1.0 0.5000000</span></span>
<span><span class="co">## chr1:11-14       C    c     NA        NA</span></span>
<span><span class="co">## chr1:1-6         A    d    2.0 1.0000000</span></span>
<span><span class="co">## chr1:7-10        B    d    0.5 0.2500000</span></span>
<span><span class="co">## chr1:11-14       C    d    1.0 1.0000000</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">summary</span><span class="op">)</span>, summary <span class="op">=</span> <span class="va">summary</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 12 ranges and 4 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand | summary.feature summary.file summary.V1.sum</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;  &lt;character&gt;      &lt;numeric&gt;</span></span>
<span><span class="co">##    [1]     chr1       1-6      * |               A            a            2.0</span></span>
<span><span class="co">##    [2]     chr1      7-10      * |               B            a            0.5</span></span>
<span><span class="co">##    [3]     chr1     11-14      * |               C            a            2.0</span></span>
<span><span class="co">##    [4]     chr1       1-6      * |               A            b            1.0</span></span>
<span><span class="co">##    [5]     chr1      7-10      * |               B            b            1.0</span></span>
<span><span class="co">##    ...      ...       ...    ... .             ...          ...            ...</span></span>
<span><span class="co">##    [8]     chr1      7-10      * |               B            c            1.0</span></span>
<span><span class="co">##    [9]     chr1     11-14      * |               C            c             NA</span></span>
<span><span class="co">##   [10]     chr1       1-6      * |               A            d            2.0</span></span>
<span><span class="co">##   [11]     chr1      7-10      * |               B            d            0.5</span></span>
<span><span class="co">##   [12]     chr1     11-14      * |               C            d            1.0</span></span>
<span><span class="co">##        summary.V1.mean</span></span>
<span><span class="co">##              &lt;numeric&gt;</span></span>
<span><span class="co">##    [1]        0.666667</span></span>
<span><span class="co">##    [2]        0.250000</span></span>
<span><span class="co">##    [3]        1.000000</span></span>
<span><span class="co">##    [4]        0.500000</span></span>
<span><span class="co">##    [5]        1.000000</span></span>
<span><span class="co">##    ...             ...</span></span>
<span><span class="co">##    [8]            0.50</span></span>
<span><span class="co">##    [9]              NA</span></span>
<span><span class="co">##   [10]            1.00</span></span>
<span><span class="co">##   [11]            0.25</span></span>
<span><span class="co">##   [12]            1.00</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="from-make_mat">From <code>make_mat</code><a class="anchor" aria-label="anchor" href="#from-make_mat"></a>
</h3>
<p><code><a href="../reference/make_mat.html">make_mat_gr()</a></code> returns a <code>GRanges</code> object for
dense matrices.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_mat.html">make_mat_gr</a></span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span>, column <span class="op">=</span> <span class="fl">4</span>, mat_name <span class="op">=</span> <span class="st">"beta"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [16:33:44.478547] [iscream::query_all] [info] Querying 3 regions from 4 bedfiles</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [16:33:44.478903] [iscream::query_all] [info] Creating metadata vectors</span></span>
<span><span class="co">## [16:33:44.478941] [iscream::query_all] [info] 7 loci found - 9993 extra rows allocated with 0 resizes</span></span>
<span><span class="co">## [16:33:44.478944] [iscream::query_all] [info] Creating dense matrix</span></span></code></pre>
<pre><code><span><span class="co">## GRanges object with 7 ranges and 4 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |         a         b         c         d</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr1         0      * |       1.0         0         0       1.0</span></span>
<span><span class="co">##   [2]     chr1         2      * |       1.0         0         1       1.0</span></span>
<span><span class="co">##   [3]     chr1         4      * |       0.0         1         0       0.0</span></span>
<span><span class="co">##   [4]     chr1         6      * |       0.0         1         0       0.0</span></span>
<span><span class="co">##   [5]     chr1         8      * |       0.5         0         1       0.5</span></span>
<span><span class="co">##   [6]     chr1        10      * |       1.0         0         0       0.0</span></span>
<span><span class="co">##   [7]     chr1        12      * |       1.0         1         0       1.0</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="summarizedexperiment">
<em><a href="https://bioconductor.org/packages/3.21/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em><a class="anchor" aria-label="anchor" href="#summarizedexperiment"></a>
</h2>
<p><code><a href="../reference/make_mat.html">make_mat_se()</a></code> returns a
<code>RangedSummarizedExperiment</code> for both sparse and dense
matrices.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">"SummarizedExperiment"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"The 'SummarizedExperiment' package must be installed for this functionality"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'MatrixGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">##     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">##     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">##     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">##     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">##     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">##     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">##     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">##     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">##     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">##     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">##     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">##     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">##     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">##     rowWeightedSds, rowWeightedVars</span></span></code></pre>
<pre><code><span><span class="co">## Welcome to Bioconductor</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Vignettes contain introductory material; view with</span></span>
<span><span class="co">##     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">##     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'Biobase'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     rowMedians</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyMissing, rowMedians</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_mat.html">make_mat_se</a></span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span>, column <span class="op">=</span> <span class="fl">4</span>, mat_name <span class="op">=</span> <span class="st">"beta"</span>, sparse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [16:33:46.697730] [iscream::query_all] [info] Querying 3 regions from 4 bedfiles</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [16:33:46.698125] [iscream::query_all] [info] Creating metadata vectors</span></span>
<span><span class="co">## [16:33:46.698152] [iscream::query_all] [info] 7 loci found - 9993 extra rows allocated with 0 resizes</span></span>
<span><span class="co">## [16:33:46.698158] [iscream::query_all] [info] Creating sparse matrix</span></span></code></pre>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 7 4 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): beta</span></span>
<span><span class="co">## rownames: NULL</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(4): a b c d</span></span>
<span><span class="co">## colData names(0):</span></span></code></pre>
<div class="section level3">
<h3 id="making-bsseq-objects">Making <code>BSseq</code> objects<a class="anchor" aria-label="anchor" href="#making-bsseq-objects"></a>
</h3>
<p>A <em><a href="https://bioconductor.org/packages/3.21/bsseq" class="external-link">bsseq</a></em>
object is a type of SummarizedExperiment, but it cannot handle sparse
matrices:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/kasperdanielhansen/bsseq" class="external-link">"bsseq"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"The 'bsseq' package must be installed for this functionality"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">mats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_mat_bsseq.html">make_mat_bsseq</a></span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span>, sparse <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [16:33:50.742104] [iscream::query_all] [info] Querying 3 regions from 4 bedfiles</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [16:33:50.742473] [iscream::query_all] [info] Creating metadata vectors</span></span>
<span><span class="co">## [16:33:50.742496] [iscream::query_all] [info] 7 loci found - 9993 extra rows allocated with 0 resizes</span></span>
<span><span class="co">## [16:33:50.742500] [iscream::query_all] [info] Creating dense matrix</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">BSseq</span>, <span class="va">mats</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of type 'BSseq' with</span></span>
<span><span class="co">##   7 methylation loci</span></span>
<span><span class="co">##   4 samples</span></span>
<span><span class="co">## has not been smoothed</span></span>
<span><span class="co">## All assays are in-memory</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] bsseq_1.44.1                SummarizedExperiment_1.38.1</span></span>
<span><span class="co">##  [3] Biobase_2.68.0              MatrixGenerics_1.20.0      </span></span>
<span><span class="co">##  [5] matrixStats_1.5.0           GenomicRanges_1.60.0       </span></span>
<span><span class="co">##  [7] GenomeInfoDb_1.44.3         IRanges_2.42.0             </span></span>
<span><span class="co">##  [9] S4Vectors_0.46.0            BiocGenerics_0.54.1        </span></span>
<span><span class="co">## [11] generics_0.1.4              iscream_0.99.7             </span></span>
<span><span class="co">## [13] BiocStyle_2.36.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] farver_2.1.2              R.utils_2.13.0           </span></span>
<span><span class="co">##  [3] Biostrings_2.76.0         bitops_1.0-9             </span></span>
<span><span class="co">##  [5] fastmap_1.2.0             RCurl_1.98-1.17          </span></span>
<span><span class="co">##  [7] GenomicAlignments_1.44.0  stringfish_0.17.0        </span></span>
<span><span class="co">##  [9] XML_3.99-0.19             digest_0.6.37            </span></span>
<span><span class="co">## [11] lifecycle_1.0.4           statmod_1.5.1            </span></span>
<span><span class="co">## [13] compiler_4.5.1            rlang_1.1.6              </span></span>
<span><span class="co">## [15] sass_0.4.10               tools_4.5.1              </span></span>
<span><span class="co">## [17] yaml_2.3.10               data.table_1.17.8        </span></span>
<span><span class="co">## [19] rtracklayer_1.68.0        knitr_1.50               </span></span>
<span><span class="co">## [21] S4Arrays_1.8.1            curl_7.0.0               </span></span>
<span><span class="co">## [23] DelayedArray_0.34.1       RColorBrewer_1.1-3       </span></span>
<span><span class="co">## [25] abind_1.4-8               BiocParallel_1.42.2      </span></span>
<span><span class="co">## [27] HDF5Array_1.36.0          R.oo_1.27.1              </span></span>
<span><span class="co">## [29] desc_1.4.3                grid_4.5.1               </span></span>
<span><span class="co">## [31] beachmat_2.24.0           Rhdf5lib_1.30.0          </span></span>
<span><span class="co">## [33] scales_1.4.0              gtools_3.9.5             </span></span>
<span><span class="co">## [35] cli_3.6.5                 rmarkdown_2.30           </span></span>
<span><span class="co">## [37] crayon_1.5.3              ragg_1.5.0               </span></span>
<span><span class="co">## [39] RcppParallel_5.1.11-1     rjson_0.2.23             </span></span>
<span><span class="co">## [41] httr_1.4.7                DelayedMatrixStats_1.30.0</span></span>
<span><span class="co">## [43] pbapply_1.7-4             cachem_1.1.0             </span></span>
<span><span class="co">## [45] rhdf5_2.52.1              parallel_4.5.1           </span></span>
<span><span class="co">## [47] BiocManager_1.30.26       XVector_0.48.0           </span></span>
<span><span class="co">## [49] restfulr_0.0.16           Matrix_1.7-3             </span></span>
<span><span class="co">## [51] jsonlite_2.0.0            bookdown_0.45            </span></span>
<span><span class="co">## [53] systemfonts_1.3.1         h5mread_1.0.1            </span></span>
<span><span class="co">## [55] locfit_1.5-9.12           limma_3.64.3             </span></span>
<span><span class="co">## [57] jquerylib_0.1.4           glue_1.8.0               </span></span>
<span><span class="co">## [59] parallelly_1.45.1         pkgdown_2.1.3            </span></span>
<span><span class="co">## [61] codetools_0.2-20          BiocIO_1.18.0            </span></span>
<span><span class="co">## [63] UCSC.utils_1.4.0          htmltools_0.5.8.1        </span></span>
<span><span class="co">## [65] rhdf5filters_1.20.0       GenomeInfoDbData_1.2.14  </span></span>
<span><span class="co">## [67] BSgenome_1.76.0           R6_2.6.1                 </span></span>
<span><span class="co">## [69] textshaping_1.0.4         sparseMatrixStats_1.20.0 </span></span>
<span><span class="co">## [71] evaluate_1.0.5            lattice_0.22-7           </span></span>
<span><span class="co">## [73] R.methodsS3_1.8.2         Rsamtools_2.24.1         </span></span>
<span><span class="co">## [75] bslib_0.9.0               Rcpp_1.1.0               </span></span>
<span><span class="co">## [77] SparseArray_1.8.1         permute_0.9-8            </span></span>
<span><span class="co">## [79] xfun_0.53                 fs_1.6.6</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by James Eapen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
