<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Improving iscream performance • iscream</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Improving iscream performance">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">iscream</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/iscream.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/performance.html">Improving iscream performance</a></li>
    <li><a class="dropdown-item" href="../articles/data_structures.html">iscream compatible data structures</a></li>
    <li><a class="dropdown-item" href="../articles/TSS.html">Plotting TSS methylation profiles</a></li>
    <li><a class="dropdown-item" href="../articles/tabix.html">iscream vs Rsamtools::scanTabix</a></li>
    <li><a class="dropdown-item" href="../articles/htslib.html">Getting htslib headers</a></li>
    <li><a class="dropdown-item" href="../articles/manuscript_data.html">Manuscript data availabiliy</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/huishenlab/iscream/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Improving iscream performance</h1>
            
            <h4 data-toc-skip class="date">10 July 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/huishenlab/iscream/blob/main/vignettes/performance.Rmd" class="external-link"><code>vignettes/performance.Rmd</code></a></small>
      <div class="d-none name"><code>performance.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>A guide to improving iscream’s runtime and memory
      efficiency</p>
    </div>
    
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<div class="section level3">
<h3 id="input-bed-files">Input BED files<a class="anchor" aria-label="anchor" href="#input-bed-files"></a>
</h3>
<p>Running this vignette requires downloading 2GB of single-cell whole
genome bisulfite sequencing (WGBS) BED files and tabix indices from this
Zenodo record: <a href="https://zenodo.org/records/14733834" class="external-link uri">https://zenodo.org/records/14733834</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"BiocFileCache"</span><span class="op">)</span></span>
<span><span class="va">cachedir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">BiocFileCache</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">snmc_zip_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcrpath</a></span><span class="op">(</span><span class="va">cachedir</span>, <span class="st">"https://zenodo.org/records/14733834/files/sc_beds.zip"</span><span class="op">)</span></span>
<span><span class="va">snmc_unzip</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"sc_beds"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html" class="external-link">unzip</a></span><span class="op">(</span><span class="va">snmc_zip_path</span>, exdir <span class="op">=</span> <span class="va">snmc_unzip</span><span class="op">)</span></span>
<span><span class="va">genes_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcrpath</a></span><span class="op">(</span><span class="va">cachedir</span>, <span class="st">"https://zenodo.org/records/14733834/files/genes.bed"</span><span class="op">)</span></span></code></pre></div>
<p>Select 100 human cell WGBS data from the snmC-seq2 <span class="citation">(<a href="#ref-luo2018a">Luo et al. 2018</a>)</span>
dataset:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snmc_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">snmc_unzip</span>, <span class="st">"sc_biscuit"</span><span class="op">)</span></span>
<span><span class="va">bedfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span></span>
<span>  <span class="va">snmc_dir</span>,</span>
<span>  pattern <span class="op">=</span> <span class="st">"*.bed.gz$"</span>,</span>
<span>  full.names <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="regions">Regions<a class="anchor" aria-label="anchor" href="#regions"></a>
</h3>
<p>Here we’ll be using 5000 gene body regions as the input:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span></span>
<span>  <span class="va">genes_file</span>,</span>
<span>  col.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="st">"start"</span>, <span class="st">"end"</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5000</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">regions</span><span class="op">)</span></span>
<span><span class="co">#&gt;       chr    start      end</span></span>
<span><span class="co">#&gt;    &lt;char&gt;    &lt;int&gt;    &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1:   chr1  1471764  1497848</span></span>
<span><span class="co">#&gt; 2:   chr1  3069167  3438621</span></span>
<span><span class="co">#&gt; 3:   chr1  2403963  2413797</span></span>
<span><span class="co">#&gt; 4:   chr1 10472287 10630758</span></span>
<span><span class="co">#&gt; 5:   chr1  2425979  2505532</span></span>
<span><span class="co">#&gt; 6:   chr1  9292893  9369532</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="multithreading">Multithreading<a class="anchor" aria-label="anchor" href="#multithreading"></a>
</h2>
<p>iscream uses multithreading to reduce querying runtime. The number of
threads can be set before or after loading the library. iscream will
read the option while loading and so can be set in
<code>.Rprofile</code></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span><span class="st">"iscream.threads"</span> <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://huishenlab.github.io/iscream/">iscream</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; iscream using 8 threads based on 'options(iscream.threads)' but parallelly::availableCores() detects 16 possibly available threads. See `?set_threads` for information on multithreading before trying to use more.</span></span></code></pre></div>
<p>Queries are parallelized across the input BED files. While this
reduces runtime, it increases memory usage as data from multiple files
are loaded into memory at the same time. For more information on
multithreading see <code>?set_threads()</code>.</p>
<div class="section level3">
<h3 id="tabix">
<code>tabix()</code><a class="anchor" aria-label="anchor" href="#tabix"></a>
</h3>
<p><code><a href="../reference/tabix.html">tabix()</a></code> queries regions from BED files much like the
<em>tabix</em> shell command, but supports multiple files and can query
them in parallel. Although fast, <code><a href="../reference/tabix.html">tabix()</a></code> may seem
unresponsive on large queries as <code><a href="https://rdrr.io/r/parallel/mclapply.html" class="external-link">parallel::mclapply()</a></code>
doesn’t show progression, but a progress bar is in development.</p>
<p>iscream has two <code><a href="../reference/tabix.html">tabix()</a></code> implementations, one using the
command line <code>tabix</code> executable and the other using the
htslib API. The command line tool is faster as it can stream to a file
which can be read from R while the htslib API stores the strings in
memory and is slower. iscream will look for the tabix executable on
startup and will only fall back to the htslip API if the executable is
not found. See <code><a href="../reference/tabix.html">?tabix</a></code> details for more information.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span></span>
<span>  <span class="va">tbx_query</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tabix.html">tabix</a></span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">regions</span>, col.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"beta"</span>, <span class="st">"coverage"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">tbx_query</span></span>
<span><span class="co">#&gt;              chr     start       end  beta coverage            file</span></span>
<span><span class="co">#&gt;           &lt;char&gt;     &lt;int&gt;     &lt;int&gt; &lt;num&gt;    &lt;int&gt;          &lt;char&gt;</span></span>
<span><span class="co">#&gt;        1:   chr1    923949    923950 0.000        1 bisc_SRR6911624</span></span>
<span><span class="co">#&gt;        2:   chr1    923953    923954 0.000        1 bisc_SRR6911624</span></span>
<span><span class="co">#&gt;        3:   chr1    923959    923960 0.000        1 bisc_SRR6911624</span></span>
<span><span class="co">#&gt;        4:   chr1    923971    923972 0.000        1 bisc_SRR6911624</span></span>
<span><span class="co">#&gt;        5:   chr1    923973    923974 0.000        1 bisc_SRR6911624</span></span>
<span><span class="co">#&gt;       ---                                                          </span></span>
<span><span class="co">#&gt; 45733375:   chr4 190179369 190179370 0.000        2 bisc_SRR6911723</span></span>
<span><span class="co">#&gt; 45733376:   chr4 190179686 190179687 1.000        1 bisc_SRR6911723</span></span>
<span><span class="co">#&gt; 45733377:   chr4 190179687 190179688 0.500        2 bisc_SRR6911723</span></span>
<span><span class="co">#&gt; 45733378:   chr4 190179753 190179754 1.000        1 bisc_SRR6911723</span></span>
<span><span class="co">#&gt; 45733379:   chr4 190179754 190179755 0.333        3 bisc_SRR6911723</span></span></code></pre></div>
<p>On 8 threads, retrieving 45,733,379 records across 100 files took
11.814 seconds.</p>
</div>
<div class="section level3">
<h3 id="summarize_regions">
<code>summarize_regions</code><a class="anchor" aria-label="anchor" href="#summarize_regions"></a>
</h3>
<p>To get a summary of the information of the gene bodies use
<code>summarize_regions</code>, providing the gene name column as the
feature column:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span></span>
<span>  <span class="va">summary_query</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_regions.html">summarize_regions</a></span><span class="op">(</span></span>
<span>    <span class="va">bedfiles</span>,</span>
<span>    <span class="va">regions</span>,</span>
<span>    columns <span class="op">=</span> <span class="fl">4</span>,</span>
<span>    col_names <span class="op">=</span> <span class="st">"beta"</span>,</span>
<span>    feature_col <span class="op">=</span> <span class="st">"gene"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; [10:51:04.214132] [iscream::summarize_regions] [info] Summarizing 5000 regions from 100 bedfiles</span></span>
<span><span class="co">#&gt; [10:51:04.214257] [iscream::summarize_regions] [info] using sum, mean, median, stddev, variance, min, max, range, count</span></span>
<span><span class="co">#&gt; [10:51:04.214269] [iscream::summarize_regions] [info] with columns 4 as beta</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">summary_query</span><span class="op">)</span></span>
<span><span class="co">#&gt;   feature            file beta.sum beta.mean beta.median beta.stddev</span></span>
<span><span class="co">#&gt; 1  ATAD3B bisc_SRR6911624   85.000 0.8585859           1   0.3502215</span></span>
<span><span class="co">#&gt; 2  PRDM16 bisc_SRR6911624  723.500 0.5288743           1   0.4976973</span></span>
<span><span class="co">#&gt; 3   PEX10 bisc_SRR6911624   15.000 0.2419355           0   0.4317514</span></span>
<span><span class="co">#&gt; 4   PEX14 bisc_SRR6911624  198.000 0.7764706           1   0.4174294</span></span>
<span><span class="co">#&gt; 5   PLCH2 bisc_SRR6911624  184.333 0.7228745           1   0.4474832</span></span>
<span><span class="co">#&gt; 6   SPSB1 bisc_SRR6911624   64.000 0.8648649           1   0.3442015</span></span>
<span><span class="co">#&gt;   beta.variance beta.min beta.max beta.range count</span></span>
<span><span class="co">#&gt; 1     0.1226551        0        1          1    99</span></span>
<span><span class="co">#&gt; 2     0.2477026        0        1          1  1368</span></span>
<span><span class="co">#&gt; 3     0.1864093        0        1          1    62</span></span>
<span><span class="co">#&gt; 4     0.1742473        0        1          1   255</span></span>
<span><span class="co">#&gt; 5     0.2002412        0        1          1   255</span></span>
<span><span class="co">#&gt; 6     0.1184746        0        1          1    74</span></span></code></pre></div>
<p>On 8 threads, summarizing the 45,733,379 records across all 100 files
took 4.948 seconds. As with <code><a href="../reference/tabix.html">tabix()</a></code>, runtime reduces and
memory usage increases with increasing thread counts.</p>
</div>
<div class="section level3">
<h3 id="make_mat">
<code>make_mat</code><a class="anchor" aria-label="anchor" href="#make_mat"></a>
</h3>
<p>The <code><a href="../reference/make_mat.html">make_mat()</a></code> function queries and stores every genomic
location within the input regions across input files. Unlike
<code><a href="../reference/summarize_regions.html">summarize_regions()</a></code> the output matrix dimensions are
unknown at runtime. Although usually quite fast, if the number of
records falling into the input regions are large and there are few
overlaps between files, this can take a long time. Here, gene bodies are
large and the final matrix can contain millions of CpG loci. Further,
with sparse data, chances are new loci are found in every file.</p>
<p>Preallocating the number of rows, however, can drastically reduce
runtime. Since we got 45 million loci/CpGs from all the BED files with
the tabix query above, we can expect approximately between 5 and 10
million unique loci as single-cell WGBS data has lower coverage than
bulk. We already have the tabix query so we can get the unique CpG count
here and use it to preallocate the matrix, reducing the number of matrix
resizes. We’ll add 100,000 extra rows on top of the existing count to be
safe since every avoided resize cuts off at least a couple seconds from
the runtime. Making tabix queries can be a relatively quick way to
approximate the CpG count of a dataset. If you haven’t done a tabix
query of the full dataset, you can approximate how many CpGs to expect
based on CpG counts in one file and the coverage of your WGBS method.
Here we make a matrix of the beta-values in the 4th column:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cpg.count</span> <span class="op">&lt;-</span> <span class="va">tbx_query</span><span class="op">$</span><span class="va">start</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">qt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="va">meth_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_mat.html">make_mat_se</a></span><span class="op">(</span></span>
<span>  <span class="va">bedfiles</span>,</span>
<span>  <span class="va">regions</span>,</span>
<span>  column <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  sparse <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  prealloc <span class="op">=</span> <span class="va">cpg.count</span> <span class="op">+</span> <span class="fl">1e5</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [10:51:11.215841] [iscream::query_all] [info] Querying 5000 regions from 100 bedfiles</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [10:51:51.998902] [iscream::query_all] [info] Creating metadata vectors</span></span>
<span><span class="co">#&gt; [10:51:52.457885] [iscream::query_all] [info] 7276107 loci found - 16250 extra rows allocated with 0 resizes</span></span>
<span><span class="co">#&gt; [10:51:57.284487] [iscream::query_all] [info] Creating sparse matrix</span></span>
<span><span class="va">meth_mat</span></span>
<span><span class="co">#&gt; class: RangedSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 7276107 100 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): value</span></span>
<span><span class="co">#&gt; rownames: NULL</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(100): bisc_SRR6911624 bisc_SRR6911625 ... bisc_SRR6911722</span></span>
<span><span class="co">#&gt;   bisc_SRR6911723</span></span>
<span><span class="co">#&gt; colData names(0):</span></span></code></pre></div>
<p>Making this 7,276,107 x 100 matrix took 50.429 seconds.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.0 (2025-04-11)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: AlmaLinux 9.6 (Sage Margay)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS/LAPACK: /usr/lib64/libopenblas-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.utf8        LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.utf8      LC_MONETARY=C.UTF-8    LC_MESSAGES=C.utf8    </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] iscream_0.99.0    data.table_1.17.6</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] crayon_1.5.3                httr_1.4.7                 </span></span>
<span><span class="co">#&gt;  [3] knitr_1.50                  xfun_0.52                  </span></span>
<span><span class="co">#&gt;  [5] UCSC.utils_1.4.0            generics_0.1.4             </span></span>
<span><span class="co">#&gt;  [7] DelayedArray_0.34.1         jsonlite_2.0.0             </span></span>
<span><span class="co">#&gt;  [9] RcppParallel_5.1.10         SummarizedExperiment_1.38.1</span></span>
<span><span class="co">#&gt; [11] S4Vectors_0.46.0            stringfish_0.16.0          </span></span>
<span><span class="co">#&gt; [13] stats4_4.5.0                MatrixGenerics_1.20.0      </span></span>
<span><span class="co">#&gt; [15] Biobase_2.68.0              grid_4.5.0                 </span></span>
<span><span class="co">#&gt; [17] abind_1.4-8                 evaluate_1.0.4             </span></span>
<span><span class="co">#&gt; [19] IRanges_2.42.0              GenomeInfoDb_1.44.0        </span></span>
<span><span class="co">#&gt; [21] compiler_4.5.0              Rcpp_1.1.0                 </span></span>
<span><span class="co">#&gt; [23] XVector_0.48.0              lattice_0.22-7             </span></span>
<span><span class="co">#&gt; [25] R6_2.6.1                    SparseArray_1.8.0          </span></span>
<span><span class="co">#&gt; [27] parallelly_1.45.0           parallel_4.5.0             </span></span>
<span><span class="co">#&gt; [29] GenomeInfoDbData_1.2.14     GenomicRanges_1.60.0       </span></span>
<span><span class="co">#&gt; [31] Matrix_1.7-3                tools_4.5.0                </span></span>
<span><span class="co">#&gt; [33] matrixStats_1.5.0           S4Arrays_1.8.1             </span></span>
<span><span class="co">#&gt; [35] BiocGenerics_0.54.0</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<!-- vim: set filetype=rmd: -->
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-luo2018a" class="csl-entry">
Luo, Chongyuan, Angeline Rivkin, Jingtian Zhou, Justin P. Sandoval,
Laurie Kurihara, Jacinta Lucero, Rosa Castanon, et al. 2018.
<span>“Robust Single-Cell <span>DNA</span> Methylome Profiling with
<span class="nocase">snmC-seq2</span>.”</span> <em>Nat Commun</em> 9
(1): 3824. <a href="https://doi.org/10.1038/s41467-018-06355-2" class="external-link">https://doi.org/10.1038/s41467-018-06355-2</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by James Eapen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
