<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Plotting TSS methylation profiles • iscream</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="../deps/MathJax-3.2.2/tex-chtml.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Plotting TSS methylation profiles">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">iscream</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/iscream.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/data_structures.html">Getting other data structures from iscream</a></li>
    <li><a class="dropdown-item" href="../articles/manuscript_data.html">Manuscript data availabiliy</a></li>
    <li><a class="dropdown-item" href="../articles/TSS.html">Plotting TSS methylation profiles</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/huishenlab/iscream/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Plotting TSS methylation profiles</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/huishenlab/iscream/blob/main/vignettes/TSS.Rmd" class="external-link"><code>vignettes/TSS.Rmd</code></a></small>
      <div class="d-none name"><code>TSS.Rmd</code></div>
    </div>

    
    
<p>The workflow here shows how <em>iscream</em> can be used to quickly
explore methylation profiles of given genomic regions.
<em>iscream’s</em> tabix querying functionality can be used to plot
methylation profiles around transcription start sites (TSS).</p>
<p><em>methscan</em> is a tool used to analyze single-cell bisulfite
sequencing data to find differentially methylated regions (DMRs) in the
genome. The plot created here is a reproduction of the TSS methylation
profile plot made in the <a href="https://anders-biostat.github.io/MethSCAn/tutorial.html#2-filtering-low-quality-cells" class="external-link"><em>methscan</em>
tutorial</a> as part of the filtering done before DMR analysis. The
<em>methscan</em> workflow to produce the plot involves three steps:</p>
<ul>
<li><p>converting the coverage BED files into Numpy sparse matrices on
disk</p></li>
<li><p>generating the TSS profiles from these matrices</p></li>
<li><p>summarizing and plotting</p></li>
</ul>
<p>The first two steps are run on the command line with
<em>methscan</em> while the third is done in R.</p>
<p>Although <em>iscream</em> is not designed to run analyses on full
genomes, it can be used to explore regions such as TSS flanking regions,
gene bodies, or DMRs found using tools like <em>methscan</em> more
efficiently. For example, using iscream, producing the plot of TSS
methylation profiles can be done in R directly from tabixed BED files
for both single-cell and bulk data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span><span class="st">"iscream.threads"</span> <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://huishenlab.github.io/iscream/">iscream</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">start_time</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>First, we generate a list of the BED file paths:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bedfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span></span>
<span>    <span class="st">"../data/methscan"</span>,</span>
<span>    pattern <span class="op">=</span> <span class="st">"*.cov.gz$"</span>,</span>
<span>    full.names <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="get-the-transcription-start-sites-and-flanking-regions">Get the Transcription start sites and flanking regions<a class="anchor" aria-label="anchor" href="#get-the-transcription-start-sites-and-flanking-regions"></a>
</h3>
<p>Then we read the provided TSS BED file and create 2kb flanking
regions around the start sites.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tss.regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/data.table/man/fread.html" class="external-link">fread</a></span><span class="op">(</span></span>
<span>    <span class="st">"../data/methscan/Mus_musculus.GRCm38.102_TSS.bed"</span>, drop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">5</span>, <span class="fl">6</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tss.regions</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="st">"tss"</span>, <span class="st">"geneID"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tss.regions</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       chr     tss             geneID</span></span>
<span><span class="co">##    &lt;char&gt;   &lt;int&gt;             &lt;char&gt;</span></span>
<span><span class="co">## 1:      1 3671498 ENSMUSG00000051951</span></span>
<span><span class="co">## 2:      1 4409241 ENSMUSG00000025900</span></span>
<span><span class="co">## 3:      1 4496413 ENSMUSG00000025902</span></span>
<span><span class="co">## 4:      1 4785739 ENSMUSG00000033845</span></span>
<span><span class="co">## 5:      1 4807823 ENSMUSG00000025903</span></span>
<span><span class="co">## 6:      1 4857814 ENSMUSG00000033813</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tss.regions</span><span class="op">[</span>, <span class="fu">`:=`</span><span class="op">(</span>tss.start <span class="op">=</span> <span class="va">tss</span> <span class="op">-</span> <span class="fl">2000</span>, tss.end <span class="op">=</span> <span class="va">tss</span> <span class="op">+</span> <span class="fl">2000</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># make a new data frame with chr, start, end as iscream requires these columns</span></span>
<span><span class="va">tss.for_query</span> <span class="op">&lt;-</span> <span class="va">tss.regions</span><span class="op">[</span>, <span class="fu">.</span><span class="op">(</span><span class="va">chr</span>, start <span class="op">=</span> <span class="va">tss.start</span>, end <span class="op">=</span> <span class="va">tss.end</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="make-a-tabix-query-of-the-tss-flanking-regions">Make a tabix query of the TSS flanking regions<a class="anchor" aria-label="anchor" href="#make-a-tabix-query-of-the-tss-flanking-regions"></a>
</h3>
<p>The <code><a href="../reference/tabix.html">tabix()</a></code> function queries the provided BED files for
the TSS flanking regions to produce a data frame:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">query_runtime.start</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">tss.query</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tabix.html">tabix</a></span><span class="op">(</span><span class="va">bedfiles</span>, <span class="va">tss.for_query</span>, aligner <span class="op">=</span> <span class="st">"bismark"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tss.query</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       chr   start     end methylation.percentage count.methylated</span></span>
<span><span class="co">##    &lt;char&gt;   &lt;num&gt;   &lt;num&gt;                  &lt;num&gt;            &lt;num&gt;</span></span>
<span><span class="co">## 1:      1 4785488 4785488                      0                0</span></span>
<span><span class="co">## 2:      1 4785513 4785513                      0                0</span></span>
<span><span class="co">## 3:      1 4785522 4785522                      0                0</span></span>
<span><span class="co">## 4:      1 4785533 4785533                      0                0</span></span>
<span><span class="co">## 5:      1 4786780 4786780                    100                1</span></span>
<span><span class="co">## 6:      1 4786886 4786886                    100                1</span></span>
<span><span class="co">##    count.unmethylated  sample</span></span>
<span><span class="co">##                 &lt;num&gt;  &lt;char&gt;</span></span>
<span><span class="co">## 1:                  2 cell_01</span></span>
<span><span class="co">## 2:                  2 cell_01</span></span>
<span><span class="co">## 3:                  2 cell_01</span></span>
<span><span class="co">## 4:                  2 cell_01</span></span>
<span><span class="co">## 5:                  0 cell_01</span></span>
<span><span class="co">## 6:                  0 cell_01</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="summarize-average-methylation-profile-around-tss">Summarize average methylation profile around TSS<a class="anchor" aria-label="anchor" href="#summarize-average-methylation-profile-around-tss"></a>
</h3>
<p>Given the CpG level methylation data frame, we now join the queried
data based on CpGs that fall within the TSS flanking regions to get the
CpGs 2kb around the TSS. We can also set a new <code>position</code>
column relative to the TSS (using rounded values as in the
<em>methscan</em> tutorial):</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># join</span></span>
<span><span class="va">tss.profile</span> <span class="op">&lt;-</span> <span class="va">tss.regions</span><span class="op">[</span><span class="va">tss.query</span>, <span class="fu">.</span><span class="op">(</span></span>
<span>    <span class="va">chr</span>,</span>
<span>    <span class="va">start</span>,</span>
<span>    position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">start</span> <span class="op">-</span> <span class="va">tss</span>, <span class="op">-</span><span class="fl">1L</span><span class="op">)</span>,</span>
<span>    <span class="va">methylation.percentage</span>,</span>
<span>    <span class="va">sample</span></span>
<span>  <span class="op">)</span>,</span>
<span>  on <span class="op">=</span> <span class="fu">.</span><span class="op">(</span><span class="va">chr</span>, <span class="va">tss.start</span> <span class="op">&lt;=</span> <span class="va">start</span>, <span class="va">tss.end</span> <span class="op">&gt;=</span> <span class="va">end</span><span class="op">)</span></span>
<span><span class="op">]</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get mean methylation by relative position and cell</span></span>
<span><span class="va">tss.summary</span> <span class="op">&lt;-</span> <span class="va">tss.profile</span><span class="op">[</span>,</span>
<span>  <span class="fu">.</span><span class="op">(</span>meth_frac <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">methylation.percentage</span><span class="op">/</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  by <span class="op">=</span> <span class="fu">.</span><span class="op">(</span><span class="va">position</span>, <span class="va">sample</span><span class="op">)</span></span>
<span><span class="op">]</span></span>
<span></span>
<span><span class="va">query_runtime</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/data.table/man/timetaken.html" class="external-link">timetaken</a></span><span class="op">(</span><span class="va">query_runtime.start</span><span class="op">)</span></span></code></pre></div>
<p>Time to make the query and compute the summary: 3.685s elapsed
(5.061s cpu).</p>
</div>
<div class="section level3">
<h3 id="plot-average-methylation-profiles-around-the-tss">Plot average methylation profiles around the TSS<a class="anchor" aria-label="anchor" href="#plot-average-methylation-profiles-around-the-tss"></a>
</h3>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tss.plot</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="va">tss.summary</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">position</span> <span class="op">/</span> <span class="fl">1000</span>, y <span class="op">=</span> <span class="va">meth_frac</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_y_continuous</span><span class="op">(</span></span>
<span>    labels<span class="op">=</span><span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent_format</a></span><span class="op">(</span>accuracy<span class="op">=</span><span class="fl">1</span><span class="op">)</span>,</span>
<span>    limits<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">.5</span>, <span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span>linewidth <span class="op">=</span> <span class="fl">.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span><span class="va">sample</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"position relative to TSS [kb]"</span>, y <span class="op">=</span> <span class="st">"DNA methylation"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">total_runtime</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/data.table/man/timetaken.html" class="external-link">timetaken</a></span><span class="op">(</span><span class="va">start_time</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tss.plot</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figure%2Ftss_plot-1.png" alt="TSS profiles"><p class="caption">
TSS profiles
</p>
</div>
<p>Total runtime, from getting the bedfiles and regions to making the
query, calculating the summaries and plotting: 3.728s elapsed (5.102s
cpu). With methscan, generating the TSS methylation profiles alone took
11 seconds.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.1 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /nix/store/p2a6x92n2gwljav85zz3hzid1nk7iq0l-blas-3/lib/libblas.so.3;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/Detroit</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] ggplot2_3.5.1      data.table_1.15.4  iscream_0.0.0.9000</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] vctrs_0.6.5        cli_3.6.3          knitr_1.47         rlang_1.1.4       </span></span>
<span><span class="co">##  [5] xfun_0.45          highr_0.11         generics_0.1.3     labeling_0.4.3    </span></span>
<span><span class="co">##  [9] RcppParallel_5.1.8 glue_1.7.0         colorspace_2.1-1   stringfish_0.16.0 </span></span>
<span><span class="co">## [13] scales_1.3.0       fansi_1.0.6        grid_4.4.1         evaluate_0.24.0   </span></span>
<span><span class="co">## [17] munsell_0.5.1      tibble_3.2.1       lifecycle_1.0.4    compiler_4.4.1    </span></span>
<span><span class="co">## [21] dplyr_1.1.4        pkgconfig_2.0.3    Rcpp_1.0.13        farver_2.1.2      </span></span>
<span><span class="co">## [25] lattice_0.22-6     R6_2.5.1           tidyselect_1.2.1   utf8_1.2.4        </span></span>
<span><span class="co">## [29] pillar_1.9.0       parallelly_1.38.0  parallel_4.4.1     magrittr_2.0.3    </span></span>
<span><span class="co">## [33] Matrix_1.7-0       withr_3.0.0        tools_4.4.1        gtable_0.3.5</span></span></code></pre>
<!-- vim: set filetype=rmd: -->
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by James Eapen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
