# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Beta value decoder
#'
#' @param encoded The bit-packed beta and cov Int
#' @return The beta value
#'
#' @keywords internal
#' @export
decode_beta <- function(encoded) {
    .Call(`_iscream_decode_beta`, encoded)
}

#' Coverage value decoder
#'
#' @param encoded The bit-packed beta and cov Int
#' @return The coverage value
#'
#' @keywords internal
#' @export
decode_cov <- function(encoded) {
    .Call(`_iscream_decode_cov`, encoded)
}

#' Get coverage values from bit-packed dense matrix
#'
#' @keywords internal
#' @export
get_cov <- function(m, nthreads) {
    invisible(.Call(`_iscream_get_cov`, m, nthreads))
}

#' Get M values from bit-packed dense matrix
#'
#' @keywords internal
#' @export
get_m <- function(m, nthreads) {
    invisible(.Call(`_iscream_get_m`, m, nthreads))
}

#' Get beta values from bit-packed dense matrix
#'
#' @keywords internal
#' @export
get_beta <- function(m, nthreads) {
    invisible(.Call(`_iscream_get_beta`, m, nthreads))
}

#' Get coverage values from bit-packed sparse matrix
#'
#' @keywords internal
#' @export
get_cov_sparse <- function(m) {
    invisible(.Call(`_iscream_get_cov_sparse`, m))
}

#' Get M values from bit-packed sparse matrix
#'
#' @keywords internal
#' @export
get_m_sparse <- function(m) {
    invisible(.Call(`_iscream_get_m_sparse`, m))
}

#' Get beta values from bit-packed sparse matrix
#'
#' @param m The bit-packed dgcMatrix
#' @return The beta value
#'
#' @keywords internal
#' @export
get_beta_sparse <- function(m) {
    invisible(.Call(`_iscream_get_beta_sparse`, m))
}

#' Get the number of available threads from OpenMP.
#'
#' This queries the number of available threads usign OpenMP, but will not
#' reliably provide an accurate available thread count. To get a more reliable
#' count that accounts for environment variables and HPC schedulers, use
#' get_threads()`. This function was pulled from
#' github.com/rdatatable/data.table
#' @param verbose Whether to be verbose on available omp threads
#'
#' @keywords internal
#' @export
#'
#' @examples
#' get_omp_threads(verbose = TRUE)
get_omp_threads <- function(verbose) {
    .Call(`_iscream_get_omp_threads`, verbose)
}

setup_logger <- function(logname = "iscream") {
    invisible(.Call(`_iscream_setup_logger`, logname))
}

#' spdlog Logging Lever Setter
#'
#' A helper function to turn a logging level given as string
#' into the current logging level
#'
#' @param name A string with the logging level. Value understood are,
#' in decreasing verbosity \sQuote{trace}, \sQuote{debug}, \sQuote{info},
#' \sQuote{warning}, \sQuote{error}, \sQuote{critical}, and \sQuote{off}.
#' Unrecognised names are equivalent to \sQuote{off}.
#' @return Nothing is returned.
#' @keywords internal
Cpp_set_log_level <- function(name) {
    invisible(.Call(`_iscream_Cpp_set_log_level`, name))
}

#' Get the current log level
#'
#' Can handle all of spdlogs levels, but iscream functions only supports
#' "info" and "debug"
#' @return The current logging level as a string
#' @export
get_log_level <- function() {
    .Call(`_iscream_get_log_level`)
}

#' Query a genomic interval from a opened htsFile and return the reads in it
#'
#' @param region Genomic region string in the form "chr:start-end"
#' @param bedFile The opened htslib bed file stream
#' @param tbx The bed file's tab-index
#' @returns A vector of strings from the matching region of the bed file
NULL

#' Get reads from multiple genomic regions from a tabixed bed file
#'
#' @param bedfile The name of the bed file - must have a corresponding tabix
#' file with the same name and .tbi extension
#' @param regions A vector of region strings in the form "chr:start-end"
NULL

#' Get htslib version and available features
#'
#' Returns the version of htslib being used by iscream and whether features
#' such as libdeflate support are available. This information may not always
#' correspond to the htslib version used during iscream's installation if a
#' different htslib version is available for linking at runtime.
#' @export
htslib_version <- function() {
    invisible(.Call(`_iscream_htslib_version`))
}

#' Query the chromosomes or seqnames from a vector of files
#' @param bedfile_vec The vector of bedfile paths
#' @return A vector of seqnames
#'
#' @keywords internal
#' @export
Cpp_query_chroms <- function(bedfile_vec, nthreads) {
    .Call(`_iscream_Cpp_query_chroms`, bedfile_vec, nthreads)
}

#' Get reads from a single genomic region from one tabixed bed file to return as CharacterVector
#'
#' @param bedfile The name of the bed file - must have a corresponding tabix
#' file with the same name and .tbi extension
#' @param regions A vector of region strings in the form "chr:start-end"
#'
#' @keywords internal
#' @export
Cpp_query_interval <- function(bedfile, regions) {
    .Call(`_iscream_Cpp_query_interval`, bedfile, regions)
}

#' Get namde list of reads from a single genomic region from one tabixed bed file
#'
#' @param bedfile The name of the bed file - must have a corresponding tabix
#' file with the same name and .tbi extension
#' @param regions A vector of region strings in the form "chr:start-end"
#'
#' @keywords internal
#' @export
scan_tabix <- function(bedfile, regions) {
    .Call(`_iscream_scan_tabix`, bedfile, regions)
}

#' Query all CpG info into M and coverage matrices
#' @param bedfiles A vector of bedfiles
#' @param regions A vector of regions
#' @param aligner The aligner used to make the WGBS BED files, only for
#' `make_bsseq_mat`
#' @param valInd The index of the data column needed for the matrix, for `make_mat`
#' @param merged Whether the input strands have been merged/collapsed
#' @param prealloc The number of rows to initialize the matrices with
#' @param nthreads Set number of threads to use overriding the
#' `"iscream.threads"` option. See `?set_threads` for more information.
#'
#' @keywords internal
#' @export
Cpp_query_all <- function(bedfiles, regions, aligner, valInd, merged, sparse, prealloc, nthreads) {
    .Call(`_iscream_Cpp_query_all`, bedfiles, regions, aligner, valInd, merged, sparse, prealloc, nthreads)
}

#' Apply a function over BED file records within genomic features
#'
#' This function should be called from `summarize_regions()` since there are few
#' sanity checks on the C++ side.
#' @param bedfiles A vector of bedfile paths
#' @param regions A vector of genomic regions
#' @param fun_vec Vector of the armadillo-supported stats functions to apply over the
#' CpGs in the ' regions: `"sum"`, `"mean"`, `"median"`, `"stddev"`,
#' `"variance"` "`count`", `"min"`,`"max"`, and `"range"`.
#' @param col_indices A vector of genomic regions
#' @param col_names A vector of genomic regions
#' @param mval Calculates M values when TRUE, use beta values when FALSE
#' @param region_rownames Whether to set rownames to the regions strings. Not
#' necessary if your regions vector is unnamed. If its names, then the "feature"
#' column is set to the names and the rownames are set to the regions string
#' @param nthreads Number of cores to use. See details.
#'
#' @details
#' The optimal number of threads depends on the number of bedfiles, but is set
#' to half the available OpenMP cores. See `?get_threads` for more details. It
#' can be manaully set with `set_threads()`.
#'
#' @keywords internal
#' @export
Cpp_summarize_regions <- function(bedfiles, regions, fun_vec, col_indices, col_names, aligner, mval = FALSE, region_rownames = FALSE, nthreads = 1L) {
    .Call(`_iscream_Cpp_summarize_regions`, bedfiles, regions, fun_vec, col_indices, col_names, aligner, mval, region_rownames, nthreads)
}

