# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' @export
}

#' Apply a function over CpGs within features
#' @param bedfiles A vector of bedfile paths
#' @param regions A vector of genomic regions
#' @param fun One of the supported functions to apply over the CpGs in the
#' regions: `"aggregate"`, `"average"`.
#' @param mval Calculates M values when TRUE, use beta values when FALSE
#' @param region_rownames Whether to set rownames to the regions strings
#' @param nthreads Number of cores to use. See details.
#'
#' @details
#' The optimal number of threads depends on the number of bedfiles, but is set
#' to half the available OpenMP cores. See `?get_threads` for more details. It
#' can be manaully set with `set_threads()`.
#'
#' @export
cpg_apply <- function(bedfiles, regions, fun, mval, region_rownames = FALSE, nthreads = 1L) {
    .Call(`_iscream_cpg_apply`, bedfiles, regions, fun, mval, region_rownames, nthreads)
}

#' Beta value decoder
#' @param encoded The bit-packed beta and cov Int
#' @return The beta value
#' @export
decode_beta <- function(encoded) {
    .Call(`_iscream_decode_beta`, encoded)
}

#' Coverage value decoder
#' @param encoded The bit-packed beta and cov Int
#' @return The coverage value
#' @export
decode_cov <- function(encoded) {
    .Call(`_iscream_decode_cov`, encoded)
}

#' M value decoder
#' @param encoded The bit-packed beta and cov Int
#' @return The M value
#' @export
decode_m <- function(encoded) {
    .Call(`_iscream_decode_m`, encoded)
}

#' Vector Decoder
#' This function unpacks the encoded \eqn{\beta} and coverage values from the
#' encoded column of a methylation matrix.
#' @param encoded An encoded sample column from a data.table methylation matrix of CpG loci by samples
#' @param measure The required decoded value: 1 for beta, 2 for coverage and 3 for M
#' @return An IntegerVector of unpacked beta, coverage or M values
#' @export
vdecoder <- function(encoded, measure) {
    .Call(`_iscream_vdecoder`, encoded, measure)
}

#' Double Vector Decoder
#' This function unpacks the encoded \eqn{\beta} and coverage values from the
#' encoded column of a methylation matrix. Same as `vdecoder`, but accepts double columns
#' @param encoded An encoded sample column from a data.table methylation matrix of CpG loci by samples
#' @param measure The required decoded value: 1 for beta, 2 for coverage and 3 for M
#' @return A DoubleVector of unpacked beta, coverage or M values
#' @export
vdouble_decoder <- function(encoded, measure) {
    .Call(`_iscream_vdouble_decoder`, encoded, measure)
}

#' Vector Encoder
#' Bit-packs a vector of beta and coverage values
#' @param beta_col Name of column containing beta values
#' @param cov_col Name of column containing coverage values
#' @export
vencoder <- function(beta_col, cov_col) {
    .Call(`_iscream_vencoder`, beta_col, cov_col)
}

#' Get the numebr of available threads
#' @param verbose Whether to be verbose on available omp threads
#' @export
get_omp_threads <- function(verbose) {
    .Call(`_iscream_get_omp_threads`, verbose)
}

#' Query a genomic interval from a opened htsFile and return the reads in it
#' @param region Genomic region string in the form "chr:start-end"
#' @param bedFile The opened htslib bed file stream
#' @param tbx The bed file's tab-index
#' @returns A vector of strings from the matching region of the bed file
NULL

#' Get reads from multiple genomic regions from a tabixed bed file
#' @param fname The name of the bed file - must have a tabix file with the same name and .tbi extension
#' @param regions A vector of regions strings of the form "chr:start-end"
NULL

#' Get reads from a single genomic region from one tabixed bed file.
#' @param fname The name of the bed file - must have a tabix file with the same name and .tbi extension
#' @param regions A vector of regions strings of the form "chr:start-end"
NULL

#' Get reads from single genomic regions from multiple tabixed bed file.
#' @param fname The name of the bed file - must have a tabix file with the same name and .tbi extension
#' @param regions A vector of regions strings of the form "chr:start-end"
#' @export
query_interval <- function(bedfiles, region) {
    .Call(`_iscream_query_interval`, bedfiles, region)
}

}

